import config

genomename = ""
if config.ref == "mm10":
    transcriptome = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-4.0.0/refdata-gex-mm10-2020-A"
    gtf = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-4.0.0/refdata-gex-mm10-2020-A/genes/genes.gtf"
    vgtf10x = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/mm10_rmsk.gtf"
    markers = "/mnt/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/mouse_gene_list.csv"
    vdj_reference = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-vdj-GRCm38-alts-ensembl-7.0.0"
    atac_reference = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-arc-mm10-2020-A-2.0.0"
    atac_reference_old = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-atac-mm10-1.2.0"
    infercnvOrder = "/mnt/ccrsf-ifx/Software/tools/inferCNV/gencode.vM9.annotation_pos2.txt"
    blacklist = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-atac-mm10-1.0.1/regions/blacklist.bed"
    sizes = "/mnt/ccrsf-ifx/RefGenomes/mm10_all/mm10.sizes"
    deeptools_bed = "/mnt/ccrsf-ifx/RefGenomes/mm10_GTF/gencode.vM20.annotation.for.deeptools.bed"
    arc_reference ="/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-arc-mm10-2020-A-2.0.0"
    genomename = "mm10"
    pipseq_reference = "/mnt/ccrsf-ifx/RefGenomes/PIPseq_REF/pipseeker-refdata-gex-mm10-2020-A"
    cytassist_v1 = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/SpacialTranscriptomic/Visium_Mouse_Transcriptome_Probe_Set_v1.0_mm10-2020-A.csv"
    cytassist_v2 = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/SpacialTranscriptomic/Visium_Mouse_Transcriptome_Probe_Set_v2.0_mm10-2020-A.csv"
if config.ref == "hg19":
    transcriptome = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-hg19-3.0.0"
    atac_reference_old = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-atac-hg19-1.2.0"
    blacklist = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-atac-hg19-1.0.1/regions/blacklist.bed"
    sizes = "/mnt/ccrsf-ifx/RefGenomes/hg19/hg19.sizes"
    deeptools_bed = "/mnt/ccrsf-ifx/RefGenomes/hg19_GTF/gencode.v19.annotation.for.deeptools.bed"
    genomename = "hg19"
if config.ref == "hg38":
    transcriptome = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-4.0.0/refdata-gex-GRCh38-2020-A"
    gtf = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-4.0.0/refdata-gex-GRCh38-2020-A/genes/genes.gtf"
    vgtf10x = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/hg38_rmsk.gtf"
    markers = "/mnt/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv"
    vdj_reference = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-vdj-GRCh38-alts-ensembl-5.0.0"
    atac_reference = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-arc-GRCh38-2020-A-2.0.0"
    atac_reference_old = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-atac-GRCh38-1.2.0"
    infercnvOrder = "/mnt/ccrsf-ifx/Software/tools/inferCNV/output_positional_gencode.v24_removedup.txt"
    blacklist = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-atac-GRCh38-1.0.1/regions/blacklist.bed"
    sizes = "/mnt/ccrsf-ifx/RefGenomes/RNA_References/hg38_30/hg38.sizes"
    deeptools_bed = "/mnt/ccrsf-ifx/RefGenomes/RNA_References/hg38_30/genes.for.deeptools.bed"
    arc_reference ="/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-arc-GRCh38-2020-A-2.0.0"
    genomename = "GRCh38"
    pipseq_reference = "/mnt/ccrsf-ifx/RefGenomes/PIPseq_REF/pipseeker-refdata-gex-GRCh38-2020-A/"
    cytassist_v1 = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/SpacialTranscriptomic/Visium_Human_Transcriptome_Probe_Set_v1.0_GRCh38-2020-A.csv"
    cytassist_v2 = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/SpacialTranscriptomic/Visium_Human_Transcriptome_Probe_Set_v2.0_GRCh38-2020-A.csv"

if config.ref == "b37":
    atac_reference_old = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-atac-b37-1.2.0"
    genomename = "b37"
if config.ref == "hg38_and_mm10":
    transcriptome = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-4.0.0/refdata-gex-GRCh38_and_mm10-2020-A"
    gtf = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-4.0.0/refdata-gex-GRCh38_and_mm10-2020-A/genes/genes.gtf"
    atac_reference = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-atac-GRCh38-and-mm10-2020-A-2.0.0"
    atac_reference_old = "/mnt/ccrsf-ifx/Software/tools/GemCode/ATAC_Ref/refdata-cellranger-atac-GRCh38-and-mm10-1.2.0"
if config.ref == "mmul":
    transcriptome = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-5.0.0/refdata-gex-Mmul_10-2020-A"
    gtf = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-cellranger-5.0.0/refdata-gex-Mmul_10-2020-A/genes/genes.gtf"
    vdj_reference = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/VDJ/vdj_IMGT_Macaca_mulatta_cellranger_7.1.0/"
if config.ref == "spol":
    transcriptome = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-gex-SpolF1D1/Spol"
    gtf = "/mnt/ccrsf-ifx/RefGenomes/10X_REF/refdata-gex-SpolF1D1/Spol_FtD1_STAR_reference.gtf"
