{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5490b4b9-3473-4587-b035-9ead6a4156bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna'"
      ],
      "text/latex": [
       "'/mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna'"
      ],
      "text/markdown": [
       "'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna'"
      ],
      "text/plain": [
       "[1] \"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "script_path = getwd()                                # %exclude_jupyterlab%\n",
    "script_path                                          # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a69fbdf1-1fe7-4f89-90ec-4fb7788ba37f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "#        '/sc_seurat_opt.R --genome=\"hg38\" --data.dir=/Volumes/ccrsf-static/singlecell_projects/DwightNissley_CS035576_6scRNASeq_101223/Analysis_CPMV/Analysis/1_B3C2_NC/outs/filtered_feature_bc_matrix --outdir=\"test_dir\"'),  # %exclude_jupyterlab%\n",
    "#       intern = T) # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3454c07e-4024-43b5-b69e-bffee79725ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "#        '/sc_seurat_opt.R --genome=\"hg38\" --data.dir=/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1C0217/DGE_filtered --outdir=\"test_dir\"'),  # %exclude_jupyterlab%\n",
    "#       intern = T) # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b07c0b85-37a4-4502-bb14-fe4e81ea171f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "        '/sc_seurat_opt.R --genome=\"hg38\" --data.dir=/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1V172/DGE_filtered/ --outdir=\"test_dir\"'),  # %exclude_jupyterlab%\n",
    "         intern = T) # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f2ac430e-9d69-4141-8602-91014330ee8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'gplots':\n",
      "  method         from \n",
      "  reorder.factor gdata\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(Matrix)\n",
    "library(MASS)\n",
    "library(dplyr)\n",
    "library(reshape2)\n",
    "library(ggplot2)\n",
    "library(URD)\n",
    "library(cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a0d9bd8a-30a8-4f2b-8293-37c97498b03c",
   "metadata": {},
   "outputs": [],
   "source": [
    "opt = readRDS(\"./test_dir/opt.rds\")                                 # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5fd741fb-787e-4dfa-bd0d-4757de96be63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$min.cells</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>$min.features</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>$data.dir</dt>\n",
       "\t\t<dd>'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1V172/DGE_filtered/'</dd>\n",
       "\t<dt>$sampleid</dt>\n",
       "\t\t<dd>'scRNA'</dd>\n",
       "\t<dt>$genome</dt>\n",
       "\t\t<dd>'hg38'</dd>\n",
       "\t<dt>$outdir</dt>\n",
       "\t\t<dd>'test_dir'</dd>\n",
       "\t<dt>$help</dt>\n",
       "\t\t<dd>FALSE</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$min.cells] 0\n",
       "\\item[\\$min.features] 0\n",
       "\\item[\\$data.dir] '/mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/split\\_pipe\\_comb/1V172/DGE\\_filtered/'\n",
       "\\item[\\$sampleid] 'scRNA'\n",
       "\\item[\\$genome] 'hg38'\n",
       "\\item[\\$outdir] 'test\\_dir'\n",
       "\\item[\\$help] FALSE\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$min.cells\n",
       ":   0\n",
       "$min.features\n",
       ":   0\n",
       "$data.dir\n",
       ":   '/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1V172/DGE_filtered/'\n",
       "$sampleid\n",
       ":   'scRNA'\n",
       "$genome\n",
       ":   'hg38'\n",
       "$outdir\n",
       ":   'test_dir'\n",
       "$help\n",
       ":   FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$min.cells\n",
       "[1] 0\n",
       "\n",
       "$min.features\n",
       "[1] 0\n",
       "\n",
       "$data.dir\n",
       "[1] \"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1V172/DGE_filtered/\"\n",
       "\n",
       "$sampleid\n",
       "[1] \"scRNA\"\n",
       "\n",
       "$genome\n",
       "[1] \"hg38\"\n",
       "\n",
       "$outdir\n",
       "[1] \"test_dir\"\n",
       "\n",
       "$help\n",
       "[1] FALSE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "opt                                                              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1a28f32b-8f51-4979-b8ff-17b80703ac5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "system(paste0(\"mkdir -p \", opt$outdir))\n",
    "setwd(opt$outdir) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "039bc4a0-e47b-4167-b9f0-8db796f4a0e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Detected ParseBio format.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Initialize variable (optional but explicit)\n",
    "count_mtx <- NULL  \n",
    "\n",
    "# Define file paths\n",
    "parsebio_files <- c(\"count_matrix.mtx\", \"all_genes.csv\", \"cell_metadata.csv\")\n",
    "tenx_files <- c(\"matrix.mtx\", \"matrix.mtx.gz\", \n",
    "                \"features.tsv\", \"features.tsv.gz\",\n",
    "                \"genes.tsv\", \"genes.tsv.gz\",\n",
    "                \"barcodes.tsv\", \"barcodes.tsv.gz\")\n",
    "\n",
    "# Check file existence\n",
    "has_parsebio <- all(file.exists(file.path(opt$data.dir, parsebio_files)))\n",
    "has_10x <- any(file.exists(file.path(opt$data.dir, tenx_files)))\n",
    "\n",
    "# Select reader\n",
    "if (has_parsebio) {\n",
    "  message(\"Detected ParseBio format.\")\n",
    "  count_mtx <- ReadParseBio(data.dir = opt$data.dir)\n",
    "} else if (has_10x) {\n",
    "  message(\"Detected 10x Genomics format.\")\n",
    "  count_mtx <- Read10X(data.dir = opt$data.dir)\n",
    "} else {\n",
    "  stop(\"No recognized data format found in: \", opt$data.dir)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "23b2f452-ec2d-44bd-88c1-75d2ef6b56e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#count_mtx <- count_mtx[(nrow(count_mtx)/2):nrow(count_mtx), 1:1000]                # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ce423a58-3c3c-4604-a474-c54cc7a764b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- CreateSeuratObject(counts = count_mtx,\n",
    "                                 min.cells = opt$min.cells,\n",
    "                                 min.features = opt$min.features,\n",
    "                                 project = opt$sampleid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ce3a0a16-5308-4a74-a3ef-94df66e0be14",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm(count_mtx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "85b328a9-7dad-4397-8943-ce934079baf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#find mitochondrial genes\n",
    "if (startsWith(opt$genome, \"mm10\") | startsWith(opt$genome, \"mm39\")) {\n",
    "        seur[[\"percent.mito\"]] <- PercentageFeatureSet(seur, pattern=\"^mt-\")\n",
    "} else if (startsWith(opt$genome, \"hg19\") | startsWith(opt$genome, \"hg38\")) {\n",
    "        seur[[\"percent.mito\"]] <- PercentageFeatureSet(seur, pattern=\"^MT-\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "459df5a7-12cb-48ba-b3d4-8f032c6cc80d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>percent.mito</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>34_01_04__s1</th><td>34</td><td>146</td><td>140</td><td>0.0000000</td></tr>\n",
       "\t<tr><th scope=row>34_01_09__s1</th><td>34</td><td>116</td><td>110</td><td>0.0000000</td></tr>\n",
       "\t<tr><th scope=row>34_01_15__s1</th><td>34</td><td> 77</td><td> 74</td><td>0.0000000</td></tr>\n",
       "\t<tr><th scope=row>34_01_34__s1</th><td>34</td><td>129</td><td>118</td><td>0.7751938</td></tr>\n",
       "\t<tr><th scope=row>34_01_37__s1</th><td>34</td><td>103</td><td> 98</td><td>0.0000000</td></tr>\n",
       "\t<tr><th scope=row>34_01_42__s1</th><td>34</td><td> 94</td><td> 87</td><td>0.0000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & orig.ident & nCount\\_RNA & nFeature\\_RNA & percent.mito\\\\\n",
       "  & <fct> & <dbl> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t34\\_01\\_04\\_\\_s1 & 34 & 146 & 140 & 0.0000000\\\\\n",
       "\t34\\_01\\_09\\_\\_s1 & 34 & 116 & 110 & 0.0000000\\\\\n",
       "\t34\\_01\\_15\\_\\_s1 & 34 &  77 &  74 & 0.0000000\\\\\n",
       "\t34\\_01\\_34\\_\\_s1 & 34 & 129 & 118 & 0.7751938\\\\\n",
       "\t34\\_01\\_37\\_\\_s1 & 34 & 103 &  98 & 0.0000000\\\\\n",
       "\t34\\_01\\_42\\_\\_s1 & 34 &  94 &  87 & 0.0000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| <!--/--> | orig.ident &lt;fct&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | percent.mito &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 34_01_04__s1 | 34 | 146 | 140 | 0.0000000 |\n",
       "| 34_01_09__s1 | 34 | 116 | 110 | 0.0000000 |\n",
       "| 34_01_15__s1 | 34 |  77 |  74 | 0.0000000 |\n",
       "| 34_01_34__s1 | 34 | 129 | 118 | 0.7751938 |\n",
       "| 34_01_37__s1 | 34 | 103 |  98 | 0.0000000 |\n",
       "| 34_01_42__s1 | 34 |  94 |  87 | 0.0000000 |\n",
       "\n"
      ],
      "text/plain": [
       "             orig.ident nCount_RNA nFeature_RNA percent.mito\n",
       "34_01_04__s1 34         146        140          0.0000000   \n",
       "34_01_09__s1 34         116        110          0.0000000   \n",
       "34_01_15__s1 34          77         74          0.0000000   \n",
       "34_01_34__s1 34         129        118          0.7751938   \n",
       "34_01_37__s1 34         103         98          0.0000000   \n",
       "34_01_42__s1 34          94         87          0.0000000   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       " 0.0000  0.0000  0.0000  0.2811  0.0000 10.0000 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>percent.mito</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 4\n",
       "\\begin{tabular}{llll}\n",
       " orig.ident & nCount\\_RNA & nFeature\\_RNA & percent.mito\\\\\n",
       " <fct> & <dbl> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 4\n",
       "\n",
       "| orig.ident &lt;fct&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | percent.mito &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     orig.ident nCount_RNA nFeature_RNA percent.mito"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(seur@meta.data)\n",
    "summary(seur@meta.data$percent.mito)                   # %exclude_jupyterlab%  \n",
    "seur@meta.data[is.na(seur@meta.data$percent.mito), ]   # %exclude_jupyterlab%  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2d74c3ce-e673-4d12-ad1b-108d2d3aa1e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "   30.0    58.0    84.0   108.2   126.0  4121.0 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "  30   31   32   33   34   35   36   37   38   39   40   41   42   43   44   45 \n",
       " 431  526  541  557  544  448  491  447  579  502  473  506  627  585  555  565 \n",
       "  46   47   48   49   50   51   52   53   54   55   56   57   58   59   60   61 \n",
       " 604  694  724  668  620  705  672  684  748  857  735  757  801  763  800  755 \n",
       "  62   63   64   65   66   67   68   69   70   71   72   73   74   75   76   77 \n",
       " 726  728  707  679  644  631  622  629  607  601  568  622  640  601  582  584 \n",
       "  78   79   80   81   82   83   84   85   86   87   88   89   90   91   92   93 \n",
       " 525  591  542  526  537  548  591  526  516  495  497  469  523  567  505  485 \n",
       "  94   95   96   97   98   99  100  101  102  103  104  105  106  107  108  109 \n",
       " 523  522  495  453  466  464  440  426  483  394  402  370  384  400  373  393 \n",
       " 110  111  112  113  114  115  116  117  118  119  120  121  122  123  124  125 \n",
       " 392  348  344  333  332  324  338  318  333  299  318  283  276  283  274  316 \n",
       " 126  127  128  129  130  131  132  133  134  135  136  137  138  139  140  141 \n",
       " 259  244  259  266  248  251  236  241  241  228  247  232  212  213  199  222 \n",
       " 142  143  144  145  146  147  148  149  150  151  152  153  154  155  156  157 \n",
       " 218  208  203  189  212  218  169  184  183  176  170  179  182  170  145  157 \n",
       " 158  159  160  161  162  163  164  165  166  167  168  169  170  171  172  173 \n",
       " 141  135  158  150  112  165  119  138  158  145  133  117  123  130  112  116 \n",
       " 174  175  176  177  178  179  180  181  182  183  184  185  186  187  188  189 \n",
       " 131  130  115  116  108   84   79  100  103   84   97  103  121   89   93   79 \n",
       " 190  191  192  193  194  195  196  197  198  199  200  201  202  203  204  205 \n",
       " 105   81   92   85   82  101   82   75   94   71   98   66   68   87   57   86 \n",
       " 206  207  208  209  210  211  212  213  214  215  216  217  218  219  220  221 \n",
       "  76   70   74   74   53   65   55   56   64   59   51   57   69   58   48   47 \n",
       " 222  223  224  225  226  227  228  229  230  231  232  233  234  235  236  237 \n",
       "  59   44   40   57   36   51   43   36   47   51   52   38   42   37   55   47 \n",
       " 238  239  240  241  242  243  244  245  246  247  248  249  250  251  252  253 \n",
       "  44   44   43   40   36   41   36   37   30   33   39   35   31   26   35   31 \n",
       " 254  255  256  257  258  259  260  261  262  263  264  265  266  267  268  269 \n",
       "  37   35   28   37   27   37   34   27   30   33   29   19   29   25   32   28 \n",
       " 270  271  272  273  274  275  276  277  278  279  280  281  282  283  284  285 \n",
       "  24   30   27   27   27   26   23   31   21   25   24   24   29   21   25   17 \n",
       " 286  287  288  289  290  291  292  293  294  295  296  297  298  299  300  301 \n",
       "  18   22   20   20   29   16   21   15   27   23    8   15   15   20   20   13 \n",
       " 302  303  304  305  306  307  308  309  310  311  312  313  314  315  316  317 \n",
       "  13   11   18   18   16   10   16   13   11   15    9   16   15   10   11   14 \n",
       " 318  319  320  321  322  323  324  325  326  327  328  329  330  331  332  333 \n",
       "  14   16   13   17   10   10   11   10    9   11   11   13   12   13   14   15 \n",
       " 334  335  336  337  338  339  340  341  342  343  344  345  346  347  348  349 \n",
       "  16   13   10   15    7    8   10   15   10   11    7   10   10   10    8   12 \n",
       " 350  351  352  353  354  355  356  357  358  359  360  361  362  363  364  365 \n",
       "   9   14   13   14    7    8   10    8    5   11    7    9    5   14    8    5 \n",
       " 366  367  368  369  370  371  372  373  374  375  376  377  378  379  380  381 \n",
       "  13    7    3   12    3   12   10    9    7    8    7    2   10    9    5    7 \n",
       " 382  383  384  385  386  387  388  389  390  391  392  393  394  395  396  397 \n",
       "  12    8    4   10    7    3    5    6    2    4    4    3    2    5    6    4 \n",
       " 398  399  400  401  402  403  404  405  406  407  408  409  410  411  412  413 \n",
       "   7    3    9    9    5    2    4    2    5    4    3    5    6    4    4    2 \n",
       " 414  415  416  417  418  419  420  421  422  423  424  425  426  427  428  429 \n",
       "   7    3    6    5    5    5    2    4    5    3    1    3    2    7    3    7 \n",
       " 430  431  432  433  434  435  436  437  438  439  440  441  442  443  444  445 \n",
       "   6    3    2    2    4    1    5    3    5    5    3    8    4    3    8   11 \n",
       " 446  447  448  449  450  451  452  453  454  456  457  458  459  460  461  462 \n",
       "   2    2    3    2    6    5    3    3    4    4    6    3    4    1    2    6 \n",
       " 463  464  465  466  467  468  469  470  471  473  474  475  476  477  478  479 \n",
       "   1    3    3    4    1    3    5    2    2    2    3    6    1    1    3    4 \n",
       " 480  481  482  483  484  486  487  489  490  491  492  493  494  495  496  497 \n",
       "   6    1    2    4    2    5    1    3    4    3    3    1    1    5    5    6 \n",
       " 498  499  500  501  502  503  504  506  507  508  509  510  511  512  513  514 \n",
       "   5    4    1    5    1    2    3    2    1    1    2    1    5    4    2    6 \n",
       " 515  516  517  518  519  520  521  524  525  526  527  528  531  532  534  535 \n",
       "   3    3    2    4    4    1    1    4    3    1    1    2    1    1    2    2 \n",
       " 536  537  539  540  542  543  545  546  547  548  549  550  551  552  553  554 \n",
       "   2    6    2    2    1    3    2    2    2    2    5    1    5    3    3    2 \n",
       " 555  556  557  558  561  562  564  565  566  567  569  570  571  572  575  576 \n",
       "   2    2    1    1    1    1    2    3    3    2    1    1    1    1    4    1 \n",
       " 577  579  580  581  583  584  585  586  587  588  591  592  594  595  596  597 \n",
       "   1    1    1    1    1    5    1    2    1    3    1    1    1    1    3    2 \n",
       " 599  600  601  602  604  605  606  607  608  609  610  611  612  613  616  617 \n",
       "   2    3    3    1    3    1    1    1    2    1    1    2    1    2    2    1 \n",
       " 618  619  621  622  623  626  627  628  629  632  634  635  636  638  639  640 \n",
       "   1    3    1    1    2    1    1    1    1    3    1    2    1    1    2    1 \n",
       " 641  644  645  646  647  649  651  653  655  657  658  659  662  663  664  665 \n",
       "   1    1    2    1    1    3    2    1    2    2    1    1    1    2    1    1 \n",
       " 667  669  671  673  674  677  678  680  683  685  687  688  689  690  692  693 \n",
       "   1    1    1    1    1    2    1    1    1    1    1    2    1    1    1    1 \n",
       " 696  697  698  703  705  706  707  709  710  711  715  716  717  718  719  720 \n",
       "   1    2    1    1    2    1    1    2    1    1    1    1    1    1    1    1 \n",
       " 721  722  724  732  734  737  738  740  742  744  745  746  747  754  757  761 \n",
       "   1    1    1    1    1    2    1    1    2    1    1    2    3    2    1    2 \n",
       " 762  764  767  769  770  771  772  777  784  789  790  795  807  810  812  819 \n",
       "   1    1    1    1    2    1    1    1    1    1    1    2    1    3    1    1 \n",
       " 821  822  823  826  827  828  830  835  836  837  838  841  842  844  845  846 \n",
       "   2    1    1    1    1    1    1    2    1    1    2    1    1    1    1    1 \n",
       " 852  856  860  862  867  869  876  880  888  889  894  897  903  905  906  909 \n",
       "   2    1    1    2    2    2    2    1    1    3    1    1    1    1    1    2 \n",
       " 916  921  922  923  926  930  938  942  945  950  951  952  955  959  962  971 \n",
       "   1    1    1    1    1    1    1    1    1    1    3    1    3    2    1    1 \n",
       " 974  979  980  982  984  990  994 1001 1008 1018 1021 1023 1024 1029 1037 1041 \n",
       "   1    1    1    1    1    1    1    2    1    1    1    2    1    1    4    1 \n",
       "1048 1053 1054 1057 1068 1081 1086 1089 1100 1104 1105 1126 1140 1141 1142 1146 \n",
       "   1    1    1    1    1    1    2    1    1    1    1    2    1    1    1    1 \n",
       "1152 1157 1158 1168 1173 1184 1189 1194 1202 1215 1217 1229 1234 1235 1237 1239 \n",
       "   1    1    1    1    1    1    1    1    1    1    1    1    1    1    1    1 \n",
       "1242 1251 1261 1268 1274 1279 1288 1293 1305 1316 1319 1320 1330 1333 1337 1340 \n",
       "   1    1    1    1    1    1    2    1    1    1    1    1    1    1    1    1 \n",
       "1341 1351 1362 1371 1390 1392 1403 1408 1430 1443 1445 1463 1464 1473 1477 1483 \n",
       "   1    1    3    1    1    1    1    1    1    1    1    1    1    1    1    1 \n",
       "1517 1525 1528 1533 1542 1566 1571 1640 1674 1703 1711 1714 1721 1791 1807 1811 \n",
       "   1    2    1    2    1    1    1    1    1    1    1    1    1    1    1    1 \n",
       "1857 1860 1912 1913 1923 2055 2079 2086 2115 2119 2127 2154 2170 2204 2221 2226 \n",
       "   1    1    1    1    1    1    1    1    1    1    1    1    1    1    1    1 \n",
       "2243 2253 2271 2272 2318 2367 2470 2475 2483 2505 2655 2667 2772 2788 3020 3032 \n",
       "   1    1    1    1    1    1    1    1    1    1    1    1    1    1    1    1 \n",
       "3173 3222 3332 3372 3610 4121 \n",
       "   1    1    1    1    1    1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "   23.0    56.0    81.0   100.8   119.0  2669.0 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "  23   25   26   27   28   29   30   31   32   33   34   35   36   37   38   39 \n",
       "   1    1    2   18   51  221  474  550  554  562  529  506  516  529  535  549 \n",
       "  40   41   42   43   44   45   46   47   48   49   50   51   52   53   54   55 \n",
       " 531  587  645  591  637  652  704  732  704  728  710  752  761  835  814  818 \n",
       "  56   57   58   59   60   61   62   63   64   65   66   67   68   69   70   71 \n",
       " 800  845  775  830  775  772  745  720  670  642  675  689  634  681  616  650 \n",
       "  72   73   74   75   76   77   78   79   80   81   82   83   84   85   86   87 \n",
       " 627  673  634  608  577  566  607  576  543  556  624  563  565  577  485  564 \n",
       "  88   89   90   91   92   93   94   95   96   97   98   99  100  101  102  103 \n",
       " 569  520  540  556  558  474  506  468  497  481  466  423  419  415  368  440 \n",
       " 104  105  106  107  108  109  110  111  112  113  114  115  116  117  118  119 \n",
       " 413  411  420  401  342  372  368  361  310  359  321  334  329  298  282  294 \n",
       " 120  121  122  123  124  125  126  127  128  129  130  131  132  133  134  135 \n",
       " 278  287  259  255  290  276  255  273  233  276  209  258  259  240  216  217 \n",
       " 136  137  138  139  140  141  142  143  144  145  146  147  148  149  150  151 \n",
       " 229  220  220  222  203  225  189  183  182  186  178  174  151  185  152  150 \n",
       " 152  153  154  155  156  157  158  159  160  161  162  163  164  165  166  167 \n",
       " 161  153  176  157  136  128  135  150  126  140  131  132  122  125  123  124 \n",
       " 168  169  170  171  172  173  174  175  176  177  178  179  180  181  182  183 \n",
       " 126  103  120  114  114  102  111  104  111   85  109  115   81   84  100   92 \n",
       " 184  185  186  187  188  189  190  191  192  193  194  195  196  197  198  199 \n",
       "  99   84   77   85   97   71   86   87   95   65   80   73   76   58   73   74 \n",
       " 200  201  202  203  204  205  206  207  208  209  210  211  212  213  214  215 \n",
       "  58   70   64   49   56   53   61   51   59   47   52   55   47   64   46   42 \n",
       " 216  217  218  219  220  221  222  223  224  225  226  227  228  229  230  231 \n",
       "  50   54   57   51   54   47   51   48   46   52   52   35   37   32   35   32 \n",
       " 232  233  234  235  236  237  238  239  240  241  242  243  244  245  246  247 \n",
       "  35   47   31   38   33   26   29   27   28   29   36   38   41   32   36   26 \n",
       " 248  249  250  251  252  253  254  255  256  257  258  259  260  261  262  263 \n",
       "  28   23   40   33   29   28   19   27   22   30   22   29   15   26   14   28 \n",
       " 264  265  266  267  268  269  270  271  272  273  274  275  276  277  278  279 \n",
       "  28   27   16   27   24   18   16   23   18   21   14   27   21   19   26   21 \n",
       " 280  281  282  283  284  285  286  287  288  289  290  291  292  293  294  295 \n",
       "  12   14   17   17   18   16   13   20   14   13   18   13   11   14    9   18 \n",
       " 296  297  298  299  300  301  302  303  304  305  306  307  308  309  310  311 \n",
       "   9   15   11   15   11    9   14   15   17   16    7   14   16   16   11   18 \n",
       " 312  313  314  315  316  317  318  319  320  321  322  323  324  325  326  327 \n",
       "  11   12   15   18    8    8   11   11   10    4   10    9    9    8   11    7 \n",
       " 328  329  330  331  332  333  334  335  336  337  338  339  340  341  342  343 \n",
       "  12    6    6   14   17   12    9    7   11   11    7    7   12    7    4    6 \n",
       " 344  345  346  347  348  349  350  351  352  353  354  355  356  357  358  359 \n",
       "   4    7   10    9    2    8    4    7   10    2    8    7    8    3   10    2 \n",
       " 360  361  362  363  364  365  366  367  368  369  370  371  372  373  374  375 \n",
       "   8    9   10    5    3    6    8    4    7    7    5    4    3    4    3    4 \n",
       " 376  377  378  379  380  381  382  383  384  385  386  387  388  389  390  391 \n",
       "   6    8    6    1    5    6    5    9    9    4    4    4    4    6    3    6 \n",
       " 392  393  394  395  396  398  399  400  401  402  403  404  405  406  407  408 \n",
       "   6    2    3    4    8    2    4    5    5    3    5    2    2    3    2    5 \n",
       " 409  410  411  412  413  414  415  416  417  418  419  420  421  422  423  425 \n",
       "   4    7    7    5    1    1    7    3    2    3    7    1    3    4    4    4 \n",
       " 426  427  428  429  430  431  432  433  434  435  436  437  438  439  440  441 \n",
       "   4    1    2    2    1    1    3    2    5    2    3    5    5    2    4    3 \n",
       " 442  443  444  445  446  447  448  449  450  451  453  454  455  456  457  458 \n",
       "   2    4    1    4    5    3    4    2    3    1    1    1    5    3    1    2 \n",
       " 459  460  461  462  463  464  465  466  467  468  469  470  471  472  473  474 \n",
       "   3    4    2    5    3    5    3    2    1    3    1    3    2    1    3    1 \n",
       " 475  476  477  478  479  481  482  483  484  485  486  487  488  489  490  491 \n",
       "   3    3    3    3    4    2    3    3    2    1    1    1    2    1    1    3 \n",
       " 492  493  494  496  498  499  500  502  503  505  506  507  508  510  512  513 \n",
       "   2    3    2    2    1    1    2    3    1    1    4    2    2    2    1    2 \n",
       " 515  516  517  518  519  520  521  523  524  525  526  527  528  531  532  533 \n",
       "   3    1    2    3    1    3    1    6    1    6    1    1    1    3    1    2 \n",
       " 534  538  539  541  542  543  545  547  551  552  554  555  556  557  560  561 \n",
       "   1    2    1    2    4    1    1    1    1    1    1    2    2    1    2    2 \n",
       " 564  565  568  569  572  573  574  575  576  577  578  579  582  584  589  590 \n",
       "   2    2    1    3    1    1    1    2    2    1    1    2    3    1    2    1 \n",
       " 591  592  593  594  595  599  603  604  606  611  612  613  614  615  618  619 \n",
       "   1    2    1    1    1    2    1    1    1    1    2    1    1    2    2    1 \n",
       " 620  621  622  627  631  633  638  640  641  642  644  645  646  648  649  650 \n",
       "   1    1    1    1    2    1    1    1    2    1    2    1    2    2    1    1 \n",
       " 651  652  653  655  657  658  662  664  666  668  669  674  677  678  679  680 \n",
       "   1    2    1    1    1    2    1    1    2    1    1    1    1    1    1    1 \n",
       " 681  682  684  687  688  689  690  695  697  698  700  701  702  703  704  705 \n",
       "   2    1    1    1    1    2    1    1    1    1    1    1    1    2    1    1 \n",
       " 706  707  708  711  716  721  727  730  731  734  736  737  740  741  742  743 \n",
       "   1    1    1    3    1    1    1    1    2    1    1    1    2    1    1    1 \n",
       " 747  749  754  755  757  759  768  772  773  774  777  779  781  786  787  796 \n",
       "   1    2    1    1    2    1    1    2    1    2    2    1    1    1    1    1 \n",
       " 802  810  815  816  817  819  821  823  825  828  829  830  835  838  842  844 \n",
       "   1    2    1    1    1    1    1    2    1    1    2    1    1    1    1    1 \n",
       " 848  851  852  854  864  866  867  870  872  880  883  894  896  902  906  908 \n",
       "   1    1    1    1    1    1    1    1    1    1    2    1    1    1    1    1 \n",
       " 919  924  925  926  929  930  945  954  959  960  964  968  973  978  982  983 \n",
       "   1    1    1    1    1    1    1    1    1    1    1    1    1    2    1    1 \n",
       " 985  988  996 1000 1005 1010 1012 1014 1016 1022 1024 1028 1037 1059 1060 1067 \n",
       "   1    2    1    1    1    2    1    1    1    1    1    1    2    1    2    1 \n",
       "1070 1078 1093 1101 1105 1115 1117 1133 1139 1140 1146 1152 1153 1158 1166 1173 \n",
       "   1    1    1    1    1    1    1    1    1    1    1    1    1    1    1    1 \n",
       "1187 1191 1198 1199 1228 1236 1278 1282 1287 1306 1338 1351 1358 1369 1374 1415 \n",
       "   1    1    1    1    1    1    1    1    1    1    1    1    1    1    1    1 \n",
       "1417 1451 1454 1488 1506 1515 1551 1554 1582 1603 1610 1620 1632 1636 1642 1648 \n",
       "   1    1    1    1    2    1    1    1    1    1    1    1    1    1    1    1 \n",
       "1682 1684 1712 1745 1755 1762 1826 1852 1890 1932 2039 2080 2084 2147 2175 2193 \n",
       "   1    1    1    1    1    1    1    1    1    1    1    1    1    1    1    1 \n",
       "2223 2400 2669 \n",
       "   1    1    1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(seur@meta.data$nCount_RNA)  # %exclude_jupyterlab%  \n",
    "table(seur@meta.data$nCount_RNA)  # %exclude_jupyterlab%\n",
    "summary(seur@meta.data$nFeature_RNA)  # %exclude_jupyterlab%  \n",
    "table(seur@meta.data$nFeature_RNA)  # %exclude_jupyterlab% "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "7bb1f235-01da-414c-98d1-91dab69e9a49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Default search for \"data\" layer in \"RNA\" assay yielded no results; utilizing \"counts\" layer instead.”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "png(\"VlnPlot_PreFilter.png\", height=7, width=7, units='in', res=200)\n",
    "VlnPlot(seur, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mito\"), ncol = 3)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5e0e0236-ac9d-45a9-9d59-c7a4cd890643",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot1 <- FeatureScatter(seur, feature1 = \"nCount_RNA\", feature2 = \"percent.mito\")\n",
    "plot2 <- FeatureScatter(seur, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\")\n",
    "png(\"FeatureScatter_PreFilter.png\", height=7, width=10, units='in', res=200)\n",
    "plot1 | plot2\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a973fb47-869d-4166-97d9-82dfd757c230",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "aa0830d0-dda4-43de-9a4e-109ac1ce0ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cS <- data.frame(libSize=seur$nCount_RNA, geneDetect=seur$nFeature_RNA)\n",
    "p_hi <- 1e-3 #p-value for filtering doublets\n",
    "p_lo <- 1e-2 #p-value for filtering poor libraries\n",
    "fitLibSize <- fitdistr(cS$libSize,\"negative binomial\")\n",
    "umi.upper.limit <- qnbinom(p_hi,size=fitLibSize$estimate[\"size\"],\n",
    "                 mu=fitLibSize$estimate[\"mu\"],lower.tail=F)\n",
    "umi.lower.limit <- qnbinom(p_lo,size=fitLibSize$estimate[\"size\"],\n",
    "                 mu=fitLibSize$estimate[\"mu\"],lower.tail=T)\n",
    "fitGeneDetect <- fitdistr(cS$geneDetect,\"negative binomial\")\n",
    "gene.upper.limit <- qnbinom(p_hi,size=fitGeneDetect$estimate[\"size\"],\n",
    "                 mu=fitGeneDetect$estimate[\"mu\"],lower.tail=F)\n",
    "gene.lower.limit <- qnbinom(p_lo,size=fitGeneDetect$estimate[\"size\"],\n",
    "                 mu=fitGeneDetect$estimate[\"mu\"],lower.tail=T)\n",
    "\n",
    "cur.mad <- mad(seur$percent.mito)\n",
    "cur.med <- median(seur$percent.mito)\n",
    "diff.val <- 4 * cur.mad\n",
    "mito.upper.limit <- cur.med + diff.val\n",
    "\n",
    "temp_doublets <- (cS$libSize > umi.upper.limit) | (cS$geneDetect > gene.upper.limit) #doublets IDed based on high library size or genes detected\n",
    "temp_crapLibs <- (cS$libSize < umi.lower.limit) | (cS$geneDetect < gene.lower.limit) #poor libraries IDed based on low library size or genes detected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "93cf47b4-79f2-49cb-8bfa-5adf89eddf5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0</li><li>10</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0\n",
       "\\item 10\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0\n",
       "2. 10\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]  0 10"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 0\n"
     ]
    }
   ],
   "source": [
    "range(seur$percent.mito)\n",
    "print(mito.upper.limit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "56260adb-ccd4-4911-bb45-170df3ce4db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter_summary <- t(data.frame(c(\"Doublets\"=sum(temp_doublets), \"Poor-Quality\"=sum(temp_crapLibs), \"Mitochondrial\"=sum(seur$percent.mito > mito.upper.limit), \"Total Filtered\"=sum(temp_doublets | temp_crapLibs | seur$percent.mito > mito.upper.limit))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "6d0bda6f-acbe-48bc-96fa-9de42ec141c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in temp_doublets | temp_crapLibs | mito_filter:\n",
      "“longer object length is not a multiple of shorter object length”\n"
     ]
    }
   ],
   "source": [
    "if (mito.upper.limit > 0) {\n",
    "  mito_filter <- seur$percent.mito > mito.upper.limit\n",
    "} else {\n",
    "  mito_filter <- rep(FALSE, length(seur$percent.mito))\n",
    "}\n",
    "\n",
    "filter_summary <- t(data.frame(c(\n",
    "  \"Doublets\"       = sum(temp_doublets),\n",
    "  \"Poor-Quality\"   = sum(temp_crapLibs),\n",
    "  \"Mitochondrial\"  = sum(mito_filter),\n",
    "  \"Total Filtered\" = sum(temp_doublets | temp_crapLibs | mito_filter)\n",
    ")))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "bf555465-d2fe-4498-b297-d7800803f159",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>38606</li><li>67122</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 38606\n",
       "\\item 67122\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 38606\n",
       "2. 67122\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 38606 67122"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(seur)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "88f8d9c0-5689-4a13-967b-6ad2bac0dba6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 1 × 4 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Doublets</th><th scope=col>Poor-Quality</th><th scope=col>Mitochondrial</th><th scope=col>Total Filtered</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>scRNA</th><td>761</td><td>0</td><td>0</td><td>761</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 1 × 4 of type int\n",
       "\\begin{tabular}{r|llll}\n",
       "  & Doublets & Poor-Quality & Mitochondrial & Total Filtered\\\\\n",
       "\\hline\n",
       "\tscRNA & 761 & 0 & 0 & 761\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 1 × 4 of type int\n",
       "\n",
       "| <!--/--> | Doublets | Poor-Quality | Mitochondrial | Total Filtered |\n",
       "|---|---|---|---|---|\n",
       "| scRNA | 761 | 0 | 0 | 761 |\n",
       "\n"
      ],
      "text/plain": [
       "      Doublets Poor-Quality Mitochondrial Total Filtered\n",
       "scRNA 761      0            0             761           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(filter_summary) = opt$sampleid\n",
    "filter_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "05ff21a7-2153-4fd2-a1d2-c2fec790e59c",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(filter_summary, 'FilterNumbers.csv', sep=',', quote=FALSE, row.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "a7bf3737-9bc8-4a65-9267-79881583c033",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- subset(seur, subset = nFeature_RNA >= gene.lower.limit & nFeature_RNA <= gene.upper.limit)\n",
    "seur <- subset(seur, subset = nCount_RNA   >= umi.lower.limit  & nCount_RNA   <= umi.upper.limit)\n",
    "# Only apply mitochondrial filtering if the threshold is positive\n",
    "if (mito.upper.limit > 0) {\n",
    "  seur <- subset(seur, subset = percent.mito <= mito.upper.limit)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "6c8c4084-3778-4084-9202-e63f00234ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# seur@misc$qc_cutoffs <- list(\n",
    "#   gene = c(lower = gene.lower.limit, upper = gene.upper.limit),\n",
    "#   umi  = c(lower = umi.lower.limit,  upper = umi.upper.limit),\n",
    "#   mito = mito.upper.limit\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "bca2bdca-c632-4271-9bca-cd0485e1e26c",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(data.frame(umi.upper.limit, umi.lower.limit, gene.upper.limit, gene.lower.limit, mito.upper.limit), \n",
    "            'FilterThresholds.csv', sep=',', quote=FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "2f831c0c-c784-46f7-b9d3-973180981e93",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Default search for \"data\" layer in \"RNA\" assay yielded no results; utilizing \"counts\" layer instead.”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "png(\"VlnPlot_Filtered.png\", height=7, width=7, units='in', res=200)\n",
    "VlnPlot(seur, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mito\"), ncol = 3)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "6375d422-815d-4eae-b487-d081fe04fee3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#new plot\n",
    "plot1 <- FeatureScatter(seur, feature1 = \"nCount_RNA\", feature2 = \"percent.mito\")\n",
    "plot2 <- FeatureScatter(seur, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\")\n",
    "png(\"FeatureScatter_PostFilter.png\", height=7, width=10, units='in', res=300)\n",
    "plot1 | plot2\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce7c3811-61da-4faf-8635-e116e12965ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- SCTransform(seur, vars.to.regress = \"percent.mito\", return.only.var.genes = FALSE, verbose = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af6ba085-bbd3-4d1e-9511-543fb66c4eed",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(seur, file = \"seur_10x_preprocessed_object.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "630a67cd-0a59-43d9-9a81-217d6ea588e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "###URD\n",
    "if (dim(seur)[[2]] > 10000) {\n",
    "  numPCs <- 50\n",
    "} else if (dim(seur)[[2]] < 500) {\n",
    "  numPCs <- 10\n",
    "} else {\n",
    "  #inputTags <- as.matrix(read.csv(expressionFile, row.names = 1))\n",
    "  mat1 <- as.matrix(GetAssayData(seur))\n",
    "  cat(\"createURD...\")\n",
    "  test <- suppressWarnings(createURD(count.data = mat1, min.cells=3, min.counts=3)) # )\n",
    "  cat(\"calcPCA...\")\n",
    "  test <- suppressWarnings(calcPCA(test, mp.factor = 2))\n",
    "  write.table(test@pca.sig,\"URD.txt\")\n",
    "  png(\"URD.png\", height=7, width=7, units='in', res=300)\n",
    "  pcSDPlot(test)\n",
    "  dev.off()\n",
    "\n",
    "  numPCs <- max(10, sum(test@pca.sig))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aefd0d39-0931-4ff6-a518-a2e8c199b6df",
   "metadata": {},
   "outputs": [],
   "source": [
    "top10 <- head(VariableFeatures(seur), 10)\n",
    "# plot variable features with and without labels\n",
    "plot1 <- VariableFeaturePlot(seur)\n",
    "plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)\n",
    "png('VariableFeatures.png', height=7, width=10, units='in', res=300)\n",
    "plot1 | plot2\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee37e668-b620-454d-b5ef-7a38f4943e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- RunPCA(seur, \n",
    "               features = VariableFeatures(seur), \n",
    "               npcs=max(20,numPCs), do.print = TRUE, \n",
    "               ndims.print = 1:5, nfeatures.print = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "420ff5cf-d169-4c8a-aa44-9d52cd9a2c51",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"VizPCAPlot.pdf\")\n",
    "for (i in  seq(1,20,2)) {\n",
    "  j = i + 1\n",
    "  print(VizDimLoadings(seur, i:j))\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b99e5645-36e3-4d68-b57a-9aee92944b5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot PCA\n",
    "pdf(\"AllPCAPlot.pdf\")\n",
    "for (i in c(1:10)) {\n",
    "  print(DimPlot(seur, dims=c(i, i+1), reduction=\"pca\"))\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "706d12dd-8c41-49bc-a06c-16f2ad697146",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PC heatmap\n",
    "pdf(\"PC_HeatmapPlot.pdf\")\n",
    "for (i in c(1:10)) {\n",
    "  DimHeatmap(seur, dims = i, cells = 500, balanced = TRUE)\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "#make PC elbow plot\n",
    "pdf(\"PC_ElbowPlot.pdf\")\n",
    "ElbowPlot(seur)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80b5abc1-dc9b-475a-a9d9-bc41ee360bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "resolutions <- c(0.1, 0.3, 0.6, 0.8)\n",
    "\n",
    "seur <- FindNeighbors(seur, dims=1:numPCs)\n",
    "seur <- RunTSNE(seur, dims=1:numPCs)\n",
    "write.csv(Embeddings(seur, reduction='tsne'), file = \"tSNECoordinates.csv\")\n",
    "seur <- RunUMAP(seur, dims=1:numPCs)\n",
    "write.csv(Embeddings(seur, reduction='umap'), file = \"UMAPCoordinates.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d705cc5-f727-4768-894b-c2a66c861365",
   "metadata": {},
   "outputs": [],
   "source": [
    "runRes <- c()\n",
    "tsnePlots <- list()\n",
    "umapPlots <- list()\n",
    "for (res in resolutions) {\n",
    "  seur <- FindClusters(seur, dims=1:numPCs, resolution = res, print.output = 0, save.SNN = T)\n",
    "  tsne <- TSNEPlot(seur) + ggtitle(paste(numPCs,\"PCs_res\", res, sep=\"\")) +\n",
    "    theme(plot.title = element_text(hjust = 0.5))\n",
    "  tsnePlots[[as.character(res)]] <- tsne\n",
    "  png(paste(\"TSNEPlotwith\",numPCs,\"PCs_\", res, \".png\", sep=\"\"), height=7, width=7, units='in', res=300)\n",
    "  print(tsne)\n",
    "  dev.off()\n",
    "  umap <- DimPlot(seur, reduction=\"umap\") + ggtitle(paste(numPCs,\"PCs_res\", res, sep=\"\")) +\n",
    "    theme(plot.title = element_text(hjust = 0.5))\n",
    "  umapPlots[[as.character(res)]] <- umap\n",
    "  png(paste(\"UMAPPlotwith\",numPCs,\"PCs_\", res, \".png\", sep=\"\"), height=7, width=7, units='in', res=300)\n",
    "  print(umap)\n",
    "  dev.off()\n",
    "\n",
    "  try({\n",
    "    seur.markers <- FindAllMarkers(object = seur, logfc.threshold = 0.25, only.pos=TRUE)\n",
    "    write.csv(seur.markers %>% group_by(cluster) %>% top_n(-100,\n",
    "                                                           p_val), paste(\"top100markers_pc\", numPCs, \"_res\", res, \".csv\", sep = \"\"))\n",
    "    saveRDS(seur.markers, paste(\"markers_res\", res, \".rds\", sep = \"\"))\n",
    "    runRes <- append(runRes, res)})\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8821c18-b955-4700-923d-1249fb550764",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save object\n",
    "saveRDS(seur, file = \"seur_10x_cluster_object.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "289836d2-00b6-474a-86a6-cd4dedb774d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"TSNEPlots.pdf\")\n",
    "for (res in tsnePlots){\n",
    "  print(res)\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46b93064-dcc1-49bc-8950-46feeb0d07b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"UMAPPlots.pdf\")\n",
    "for (res in umapPlots){\n",
    "  print(res)\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f1a272a-5ccb-4e68-a5ab-b4611a38f58f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute theoretical maximum size allowed by dist()\n",
    "INT_MAX <- .Machine$integer.max                 # ~2,147,483,647\n",
    "n_max   <- floor((1 + sqrt(1 + 8*INT_MAX))/2)   # ~65536\n",
    "threshold_n <- 100 * as.integer(n_max / 100)    # 65500, rounded for safety\n",
    "\n",
    "cat(\"Auto subsetting if cell number >\", threshold_n, \"\\n\")\n",
    "# Subset automatically if you exceed that limit\n",
    "if (ncol(seur) > threshold_n) {\n",
    "  set.seed(1)\n",
    "  cells.keep <- sample(colnames(seur), threshold_n)\n",
    "  seur <- subset(seur, cells = cells.keep)\n",
    "  cat(\"Subsetting applied:\", length(cells.keep), \"cells kept.\\n\")\n",
    "} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01fdc8f6-9ad8-40e5-b6ca-85ff18d7dda2",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Create Silhoutte Plots\n",
    "for (res in runRes){\n",
    "  coord <- Embeddings(seur, reduction.type='pca')[,1:numPCs]\n",
    "  Idents(seur) <- seur@meta.data[[paste0('SCT_snn_res.', res)]]\n",
    "  clusters <- Idents(seur)\n",
    "  d <- dist(coord, method=\"euclidean\")\n",
    "  sil<-silhouette(as.numeric(clusters), dist=d)\n",
    "  #silPlot <- recordPlot()\n",
    "  pdf(paste0(\"SilhouettePlot_res\",res,\".pdf\"))#, height=7, width=7, units='in', res=300)\n",
    "  plot(sil, col=as.factor(clusters[order(clusters, decreasing=FALSE)]), main=paste(\"Silhouette plot of Seurat clustering - resolution \", res, sep=\"\"), lty=2)\n",
    "  abline(v=mean(sil[,3]), col=\"red4\", lty=2)\n",
    "  dev.off()\n",
    "}\n",
    "\n",
    "##Remove resolutions that failed marker generation\n",
    "print(runRes)\n",
    "for (res in setdiff(resolutions, runRes)){\n",
    "  seur@meta.data[paste0('SCT_snn_res.', res)] <- NULL\n",
    "}\n",
    "\n",
    "write(min(runRes), \"minRes.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdc254a4-7f34-47b5-8fb4-e48b75be85c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af756324-3533-445b-ab14-b1b3952b6b84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'jupyter nbconvert --to script --output sc_seurat /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.ipynb&gt; /dev/null 2&gt;&amp;1 '"
      ],
      "text/latex": [
       "'jupyter nbconvert --to script --output sc\\_seurat /mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna/sc\\_seurat.ipynb> /dev/null 2>\\&1 '"
      ],
      "text/markdown": [
       "'jupyter nbconvert --to script --output sc_seurat /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.ipynb&gt; /dev/null 2&gt;&amp;1 '"
      ],
      "text/plain": [
       "[1] \"jupyter nbconvert --to script --output sc_seurat /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.ipynb> /dev/null 2>&1 \""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "notebook_prefix = \"sc_seurat\"                                                         # %exclude_jupyterlab%\n",
    "notebook_name = paste0(notebook_prefix, \".ipynb\")                                     # %exclude_jupyterlab%\n",
    "notebook_r = paste0(script_path, \"/\", paste0(notebook_prefix, \".r\"))                  # %exclude_jupyterlab%\n",
    "notebook_path = paste0(script_path, \"/\", notebook_name)                               # %exclude_jupyterlab%\n",
    "opt_name = paste0(script_path, \"/\", sub(\".ipynb\", \"_opt.R\", notebook_name))           # %exclude_jupyterlab%\n",
    "output = paste0(script_path, \"/\", sub(\".ipynb\", \".prod.R\", notebook_name))            # %exclude_jupyterlab%\n",
    "cmd1 = paste0(\"jupyter nbconvert --to script --output \",                              # %exclude_jupyterlab%\n",
    "             notebook_prefix, ' ', notebook_path, \"> /dev/null 2>&1 \")                # %exclude_jupyterlab%\n",
    "cmd1                                                                                  # %exclude_jupyterlab%\n",
    "system(cmd1, intern = TRUE)                                                           # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d1c24963-ff07-4aa9-abe9-cea08fc9e370",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'cat -s /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat_opt.R /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.r |grep -v exclude_jupyterlab &gt; /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.prod.R 2&gt;&amp;1'"
      ],
      "text/latex": [
       "'cat -s /mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna/sc\\_seurat\\_opt.R /mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna/sc\\_seurat.r \\textbar{}grep -v exclude\\_jupyterlab > /mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna/sc\\_seurat.prod.R 2>\\&1'"
      ],
      "text/markdown": [
       "'cat -s /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat_opt.R /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.r |grep -v exclude_jupyterlab &gt; /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.prod.R 2&gt;&amp;1'"
      ],
      "text/plain": [
       "[1] \"cat -s /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat_opt.R /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.r |grep -v exclude_jupyterlab > /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.prod.R 2>&1\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cmd2 = paste0('cat -s ', opt_name, ' ', notebook_r,                                   # %exclude_jupyterlab%\n",
    "             ' |grep -v exclude_jupyterlab > ', output,  ' 2>&1')                     # %exclude_jupyterlab%\n",
    "cmd2                                                                                  # %exclude_jupyterlab%\n",
    "system(cmd2, intern = T)                                                              # %exclude_jupyterlab%\n",
    "system(paste0(\"rm \", notebook_r))                                                     # %exclude_jupyterlab%  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b366482-e5ba-487b-a0f1-c982db2f19f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "531b32f1-3e03-4325-b0df-8d078527d584",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
