{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5490b4b9-3473-4587-b035-9ead6a4156bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna'"
      ],
      "text/latex": [
       "'/mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna'"
      ],
      "text/markdown": [
       "'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna'"
      ],
      "text/plain": [
       "[1] \"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "script_path = getwd()                                # %exclude_jupyterlab%\n",
    "script_path                                          # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a69fbdf1-1fe7-4f89-90ec-4fb7788ba37f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "#        '/sc_seurat_opt.R --genome=\"hg38\" --data.dir=/Volumes/ccrsf-static/singlecell_projects/DwightNissley_CS035576_6scRNASeq_101223/Analysis_CPMV/Analysis/1_B3C2_NC/outs/filtered_feature_bc_matrix --outdir=\"test_dir\"'),  # %exclude_jupyterlab%\n",
    "#       intern = T) # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3454c07e-4024-43b5-b69e-bffee79725ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "#        '/sc_seurat_opt.R --genome=\"hg38\" --data.dir=/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1C0217/DGE_filtered --outdir=\"test_dir\"'),  # %exclude_jupyterlab%\n",
    "#       intern = T) # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b07c0b85-37a4-4502-bb14-fe4e81ea171f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "        '/sc_seurat_opt.R --genome=\"hg38\" --data.dir=/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1V172/DGE_filtered/ --outdir=\"test_dir\"'),  # %exclude_jupyterlab%\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f2ac430e-9d69-4141-8602-91014330ee8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'gplots':\n",
      "  method         from \n",
      "  reorder.factor gdata\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(Matrix)\n",
    "library(MASS)\n",
    "library(dplyr)\n",
    "library(reshape2)\n",
    "library(ggplot2)\n",
    "library(URD)\n",
    "library(cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a0d9bd8a-30a8-4f2b-8293-37c97498b03c",
   "metadata": {},
   "outputs": [],
   "source": [
    "opt = readRDS(\"./test_dir/opt.rds\")                                 # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5fd741fb-787e-4dfa-bd0d-4757de96be63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$min.cells</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>$min.features</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>$data.dir</dt>\n",
       "\t\t<dd>'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1V172/DGE_filtered/'</dd>\n",
       "\t<dt>$sampleid</dt>\n",
       "\t\t<dd>'scRNA'</dd>\n",
       "\t<dt>$genome</dt>\n",
       "\t\t<dd>'hg38'</dd>\n",
       "\t<dt>$outdir</dt>\n",
       "\t\t<dd>'test_dir'</dd>\n",
       "\t<dt>$help</dt>\n",
       "\t\t<dd>FALSE</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$min.cells] 1\n",
       "\\item[\\$min.features] 1\n",
       "\\item[\\$data.dir] '/mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/split\\_pipe\\_comb/1V172/DGE\\_filtered/'\n",
       "\\item[\\$sampleid] 'scRNA'\n",
       "\\item[\\$genome] 'hg38'\n",
       "\\item[\\$outdir] 'test\\_dir'\n",
       "\\item[\\$help] FALSE\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$min.cells\n",
       ":   1\n",
       "$min.features\n",
       ":   1\n",
       "$data.dir\n",
       ":   '/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1V172/DGE_filtered/'\n",
       "$sampleid\n",
       ":   'scRNA'\n",
       "$genome\n",
       ":   'hg38'\n",
       "$outdir\n",
       ":   'test_dir'\n",
       "$help\n",
       ":   FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$min.cells\n",
       "[1] 1\n",
       "\n",
       "$min.features\n",
       "[1] 1\n",
       "\n",
       "$data.dir\n",
       "[1] \"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1V172/DGE_filtered/\"\n",
       "\n",
       "$sampleid\n",
       "[1] \"scRNA\"\n",
       "\n",
       "$genome\n",
       "[1] \"hg38\"\n",
       "\n",
       "$outdir\n",
       "[1] \"test_dir\"\n",
       "\n",
       "$help\n",
       "[1] FALSE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "opt                                                              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1a28f32b-8f51-4979-b8ff-17b80703ac5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "system(paste0(\"mkdir -p \", opt$outdir))\n",
    "setwd(opt$outdir) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "039bc4a0-e47b-4167-b9f0-8db796f4a0e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Detected ParseBio format.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Initialize variable (optional but explicit)\n",
    "count_mtx <- NULL  \n",
    "\n",
    "# Define file paths\n",
    "parsebio_files <- c(\"count_matrix.mtx\", \"all_genes.csv\", \"cell_metadata.csv\")\n",
    "tenx_files <- c(\"matrix.mtx\", \"matrix.mtx.gz\", \n",
    "                \"features.tsv\", \"features.tsv.gz\",\n",
    "                \"genes.tsv\", \"genes.tsv.gz\",\n",
    "                \"barcodes.tsv\", \"barcodes.tsv.gz\")\n",
    "\n",
    "# Check file existence\n",
    "has_parsebio <- all(file.exists(file.path(opt$data.dir, parsebio_files)))\n",
    "has_10x <- any(file.exists(file.path(opt$data.dir, tenx_files)))\n",
    "\n",
    "# Select reader\n",
    "if (has_parsebio) {\n",
    "  message(\"Detected ParseBio format.\")\n",
    "  count_mtx <- ReadParseBio(data.dir = opt$data.dir)\n",
    "} else if (has_10x) {\n",
    "  message(\"Detected 10x Genomics format.\")\n",
    "  count_mtx <- Read10X(data.dir = opt$data.dir)\n",
    "} else {\n",
    "  stop(\"No recognized data format found in: \", opt$data.dir)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23b2f452-ec2d-44bd-88c1-75d2ef6b56e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_mtx <- count_mtx[(nrow(count_mtx)/2):nrow(count_mtx), 1:1000]                # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ce423a58-3c3c-4604-a474-c54cc7a764b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- CreateSeuratObject(counts = count_mtx,\n",
    "                                 min.cells = opt$min.cells,\n",
    "                                 min.features = opt$min.features,\n",
    "                                 project = opt$sampleid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ce3a0a16-5308-4a74-a3ef-94df66e0be14",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm(count_mtx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "85b328a9-7dad-4397-8943-ce934079baf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#find mitochondrial genes\n",
    "if (startsWith(opt$genome, \"mm10\") | startsWith(opt$genome, \"mm39\")) {\n",
    "        seur[[\"percent.mito\"]] <- PercentageFeatureSet(seur, pattern=\"^mt-\")\n",
    "} else if (startsWith(opt$genome, \"hg19\") | startsWith(opt$genome, \"hg38\")) {\n",
    "        seur[[\"percent.mito\"]] <- PercentageFeatureSet(seur, pattern=\"^MT-\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "459df5a7-12cb-48ba-b3d4-8f032c6cc80d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>percent.mito</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>34_01_04__s1</th><td>34</td><td>14</td><td>14</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>34_01_09__s1</th><td>34</td><td> 7</td><td> 4</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>34_01_15__s1</th><td>34</td><td> 9</td><td> 6</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>34_01_34__s1</th><td>34</td><td>16</td><td>10</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>34_01_37__s1</th><td>34</td><td> 6</td><td> 4</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>34_01_42__s1</th><td>34</td><td> 8</td><td> 6</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & orig.ident & nCount\\_RNA & nFeature\\_RNA & percent.mito\\\\\n",
       "  & <fct> & <dbl> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t34\\_01\\_04\\_\\_s1 & 34 & 14 & 14 & 0\\\\\n",
       "\t34\\_01\\_09\\_\\_s1 & 34 &  7 &  4 & 0\\\\\n",
       "\t34\\_01\\_15\\_\\_s1 & 34 &  9 &  6 & 0\\\\\n",
       "\t34\\_01\\_34\\_\\_s1 & 34 & 16 & 10 & 0\\\\\n",
       "\t34\\_01\\_37\\_\\_s1 & 34 &  6 &  4 & 0\\\\\n",
       "\t34\\_01\\_42\\_\\_s1 & 34 &  8 &  6 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| <!--/--> | orig.ident &lt;fct&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | percent.mito &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 34_01_04__s1 | 34 | 14 | 14 | 0 |\n",
       "| 34_01_09__s1 | 34 |  7 |  4 | 0 |\n",
       "| 34_01_15__s1 | 34 |  9 |  6 | 0 |\n",
       "| 34_01_34__s1 | 34 | 16 | 10 | 0 |\n",
       "| 34_01_37__s1 | 34 |  6 |  4 | 0 |\n",
       "| 34_01_42__s1 | 34 |  8 |  6 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "             orig.ident nCount_RNA nFeature_RNA percent.mito\n",
       "34_01_04__s1 34         14         14           0           \n",
       "34_01_09__s1 34          7          4           0           \n",
       "34_01_15__s1 34          9          6           0           \n",
       "34_01_34__s1 34         16         10           0           \n",
       "34_01_37__s1 34          6          4           0           \n",
       "34_01_42__s1 34          8          6           0           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. \n",
       " 0.00000  0.00000  0.00000  0.02704  0.00000 16.66667 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>percent.mito</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 4\n",
       "\\begin{tabular}{llll}\n",
       " orig.ident & nCount\\_RNA & nFeature\\_RNA & percent.mito\\\\\n",
       " <fct> & <dbl> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 4\n",
       "\n",
       "| orig.ident &lt;fct&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | percent.mito &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     orig.ident nCount_RNA nFeature_RNA percent.mito"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(seur@meta.data)\n",
    "summary(seur@meta.data$percent.mito)                   # %exclude_jupyterlab%  \n",
    "seur@meta.data[is.na(seur@meta.data$percent.mito), ]   # %exclude_jupyterlab%  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2d74c3ce-e673-4d12-ad1b-108d2d3aa1e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "   1.00    6.00    8.00   10.22   12.00  332.00 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20 \n",
       "  9  27  64  67  82  90 103  84  83  75  60  41  36  37  23  17  16   8  10   9 \n",
       " 21  22  23  24  25  26  27  29  31  32  33  35  36  39  47  49  52  56  80 196 \n",
       "  9   5   9   6   6   3   1   2   1   3   2   1   1   1   1   1   1   1   1   1 \n",
       "232 332 \n",
       "  1   1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "   1.00    4.00    6.00    7.55    9.00  149.00 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20 \n",
       " 12  58  92 110 128 122 114  93  70  39  39  30  14  14  14   8   3   7   2   7 \n",
       " 21  22  23  24  25  26  27  28  30  31  36  37  38  42 104 110 149 \n",
       "  5   1   2   2   1   1   1   1   1   1   1   1   1   1   1   1   1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(seur@meta.data$nCount_RNA)  # %exclude_jupyterlab%  \n",
    "table(seur@meta.data$nCount_RNA)  # %exclude_jupyterlab%\n",
    "summary(seur@meta.data$nFeature_RNA)  # %exclude_jupyterlab%  \n",
    "table(seur@meta.data$nFeature_RNA)  # %exclude_jupyterlab% "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7bb1f235-01da-414c-98d1-91dab69e9a49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Default search for \"data\" layer in \"RNA\" assay yielded no results; utilizing \"counts\" layer instead.”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "png(\"VlnPlot_PreFilter.png\", height=7, width=7, units='in', res=200)\n",
    "VlnPlot(seur, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mito\"), ncol = 3)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5e0e0236-ac9d-45a9-9d59-c7a4cd890643",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot1 <- FeatureScatter(seur, feature1 = \"nCount_RNA\", feature2 = \"percent.mito\")\n",
    "plot2 <- FeatureScatter(seur, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\")\n",
    "png(\"FeatureScatter_PreFilter.png\", height=7, width=10, units='in', res=200)\n",
    "plot1 | plot2\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a973fb47-869d-4166-97d9-82dfd757c230",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "aa0830d0-dda4-43de-9a4e-109ac1ce0ccd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in densfun(x, parm[1], parm[2], ...):\n",
      "“NaNs produced”\n",
      "Warning message in densfun(x, parm[1], parm[2], ...):\n",
      "“NaNs produced”\n",
      "Warning message in densfun(x, parm[1], parm[2], ...):\n",
      "“NaNs produced”\n",
      "Warning message in densfun(x, parm[1], parm[2], ...):\n",
      "“NaNs produced”\n",
      "Warning message in densfun(x, parm[1], parm[2], ...):\n",
      "“NaNs produced”\n",
      "Warning message in densfun(x, parm[1], parm[2], ...):\n",
      "“NaNs produced”\n"
     ]
    }
   ],
   "source": [
    "cS <- data.frame(libSize=seur$nCount_RNA, geneDetect=seur$nFeature_RNA)\n",
    "p_hi <- 1e-3 #p-value for filtering doublets\n",
    "p_lo <- 1e-2 #p-value for filtering poor libraries\n",
    "fitLibSize <- fitdistr(cS$libSize,\"negative binomial\")\n",
    "umi.upper.limit <- qnbinom(p_hi,size=fitLibSize$estimate[\"size\"],\n",
    "                 mu=fitLibSize$estimate[\"mu\"],lower.tail=F)\n",
    "umi.lower.limit <- qnbinom(p_lo,size=fitLibSize$estimate[\"size\"],\n",
    "                 mu=fitLibSize$estimate[\"mu\"],lower.tail=T)\n",
    "fitGeneDetect <- fitdistr(cS$geneDetect,\"negative binomial\")\n",
    "gene.upper.limit <- qnbinom(p_hi,size=fitGeneDetect$estimate[\"size\"],\n",
    "                 mu=fitGeneDetect$estimate[\"mu\"],lower.tail=F)\n",
    "gene.lower.limit <- qnbinom(p_lo,size=fitGeneDetect$estimate[\"size\"],\n",
    "                 mu=fitGeneDetect$estimate[\"mu\"],lower.tail=T)\n",
    "\n",
    "cur.mad <- mad(seur$percent.mito)\n",
    "cur.med <- median(seur$percent.mito)\n",
    "diff.val <- 4 * cur.mad\n",
    "mito.upper.limit <- cur.med + diff.val\n",
    "\n",
    "temp_doublets <- (cS$libSize > umi.upper.limit) | (cS$geneDetect > gene.upper.limit) #doublets IDed based on high library size or genes detected\n",
    "temp_crapLibs <- (cS$libSize < umi.lower.limit) | (cS$geneDetect < gene.lower.limit) #poor libraries IDed based on low library size or genes detected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "93cf47b4-79f2-49cb-8bfa-5adf89eddf5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 0\n"
     ]
    }
   ],
   "source": [
    "print(mito.upper.limit)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "56260adb-ccd4-4911-bb45-170df3ce4db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter_summary <- t(data.frame(c(\"Doublets\"=sum(temp_doublets), \"Poor-Quality\"=sum(temp_crapLibs), \"Mitochondrial\"=sum(seur$percent.mito > mito.upper.limit), \"Total Filtered\"=sum(temp_doublets | temp_crapLibs | seur$percent.mito > mito.upper.limit))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6d0bda6f-acbe-48bc-96fa-9de42ec141c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_summary <- t(data.frame(c(\n",
    "  \"Doublets\"=sum(temp_doublets),\n",
    "  \"Poor-Quality\"=sum(temp_crapLibs),\n",
    "  \"Mitochondrial\"=sum(seur$percent.mito >= mito.upper.limit),\n",
    "  \"Total Filtered\"=sum(temp_doublets | temp_crapLibs | seur$percent.mito >= mito.upper.limit)\n",
    ")))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "bf555465-d2fe-4498-b297-d7800803f159",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>2520</li><li>999</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 2520\n",
       "\\item 999\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 2520\n",
       "2. 999\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 2520  999"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(seur)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "88f8d9c0-5689-4a13-967b-6ad2bac0dba6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 1 × 4 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Doublets</th><th scope=col>Poor-Quality</th><th scope=col>Mitochondrial</th><th scope=col>Total Filtered</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>scRNA</th><td>10</td><td>0</td><td>999</td><td>999</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 1 × 4 of type int\n",
       "\\begin{tabular}{r|llll}\n",
       "  & Doublets & Poor-Quality & Mitochondrial & Total Filtered\\\\\n",
       "\\hline\n",
       "\tscRNA & 10 & 0 & 999 & 999\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 1 × 4 of type int\n",
       "\n",
       "| <!--/--> | Doublets | Poor-Quality | Mitochondrial | Total Filtered |\n",
       "|---|---|---|---|---|\n",
       "| scRNA | 10 | 0 | 999 | 999 |\n",
       "\n"
      ],
      "text/plain": [
       "      Doublets Poor-Quality Mitochondrial Total Filtered\n",
       "scRNA 10       0            999           999           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(filter_summary) = opt$sampleid\n",
    "filter_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "05ff21a7-2153-4fd2-a1d2-c2fec790e59c",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(filter_summary, 'FilterNumbers.csv', sep=',', quote=FALSE, row.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a7bf3737-9bc8-4a65-9267-79881583c033",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- subset(seur, subset = nFeature_RNA >= gene.lower.limit & nFeature_RNA <= gene.upper.limit)\n",
    "seur <- subset(seur, subset = nCount_RNA   >= umi.lower.limit  & nCount_RNA   <= umi.upper.limit)\n",
    "# Only apply mitochondrial filtering if the threshold is positive\n",
    "if (mito.upper.limit > 0) {\n",
    "  seur <- subset(seur, subset = percent.mito <= mito.upper.limit)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c8c4084-3778-4084-9202-e63f00234ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# seur@misc$qc_cutoffs <- list(\n",
    "#   gene = c(lower = gene.lower.limit, upper = gene.upper.limit),\n",
    "#   umi  = c(lower = umi.lower.limit,  upper = umi.upper.limit),\n",
    "#   mito = mito.upper.limit\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bca2bdca-c632-4271-9bca-cd0485e1e26c",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(data.frame(umi.upper.limit, umi.lower.limit, gene.upper.limit, gene.lower.limit, mito.upper.limit), \n",
    "            'FilterThresholds.csv', sep=',', quote=FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f831c0c-c784-46f7-b9d3-973180981e93",
   "metadata": {},
   "outputs": [],
   "source": [
    "png(\"VlnPlot_Filtered.png\", height=7, width=7, units='in', res=200)\n",
    "VlnPlot(seur, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mito\"), ncol = 3)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6375d422-815d-4eae-b487-d081fe04fee3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#new plot\n",
    "plot1 <- FeatureScatter(seur, feature1 = \"nCount_RNA\", feature2 = \"percent.mito\")\n",
    "plot2 <- FeatureScatter(seur, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\")\n",
    "png(\"FeatureScatter_PostFilter.png\", height=7, width=10, units='in', res=300)\n",
    "plot1 | plot2\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce7c3811-61da-4faf-8635-e116e12965ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- SCTransform(seur, vars.to.regress = \"percent.mito\", return.only.var.genes = FALSE, verbose = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af6ba085-bbd3-4d1e-9511-543fb66c4eed",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(seur, file = \"seur_10x_preprocessed_object.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "630a67cd-0a59-43d9-9a81-217d6ea588e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "###URD\n",
    "if (dim(seur)[[2]] > 10000) {\n",
    "  numPCs <- 50\n",
    "} else if (dim(seur)[[2]] < 500) {\n",
    "  numPCs <- 10\n",
    "} else {\n",
    "  #inputTags <- as.matrix(read.csv(expressionFile, row.names = 1))\n",
    "  mat1 <- as.matrix(GetAssayData(seur))\n",
    "  cat(\"createURD...\")\n",
    "  test <- suppressWarnings(createURD(count.data = mat1, min.cells=3, min.counts=3)) # )\n",
    "  cat(\"calcPCA...\")\n",
    "  test <- suppressWarnings(calcPCA(test, mp.factor = 2))\n",
    "  write.table(test@pca.sig,\"URD.txt\")\n",
    "  png(\"URD.png\", height=7, width=7, units='in', res=300)\n",
    "  pcSDPlot(test)\n",
    "  dev.off()\n",
    "\n",
    "  numPCs <- max(10, sum(test@pca.sig))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aefd0d39-0931-4ff6-a518-a2e8c199b6df",
   "metadata": {},
   "outputs": [],
   "source": [
    "top10 <- head(VariableFeatures(seur), 10)\n",
    "# plot variable features with and without labels\n",
    "plot1 <- VariableFeaturePlot(seur)\n",
    "plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)\n",
    "png('VariableFeatures.png', height=7, width=10, units='in', res=300)\n",
    "plot1 | plot2\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee37e668-b620-454d-b5ef-7a38f4943e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- RunPCA(seur, \n",
    "               features = VariableFeatures(seur), \n",
    "               npcs=max(20,numPCs), do.print = TRUE, \n",
    "               ndims.print = 1:5, nfeatures.print = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "420ff5cf-d169-4c8a-aa44-9d52cd9a2c51",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"VizPCAPlot.pdf\")\n",
    "for (i in  seq(1,20,2)) {\n",
    "  j = i + 1\n",
    "  print(VizDimLoadings(seur, i:j))\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b99e5645-36e3-4d68-b57a-9aee92944b5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot PCA\n",
    "pdf(\"AllPCAPlot.pdf\")\n",
    "for (i in c(1:10)) {\n",
    "  print(DimPlot(seur, dims=c(i, i+1), reduction=\"pca\"))\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "706d12dd-8c41-49bc-a06c-16f2ad697146",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PC heatmap\n",
    "pdf(\"PC_HeatmapPlot.pdf\")\n",
    "for (i in c(1:10)) {\n",
    "  DimHeatmap(seur, dims = i, cells = 500, balanced = TRUE)\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "#make PC elbow plot\n",
    "pdf(\"PC_ElbowPlot.pdf\")\n",
    "ElbowPlot(seur)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80b5abc1-dc9b-475a-a9d9-bc41ee360bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "resolutions <- c(0.1, 0.3, 0.6, 0.8)\n",
    "\n",
    "seur <- FindNeighbors(seur, dims=1:numPCs)\n",
    "seur <- RunTSNE(seur, dims=1:numPCs)\n",
    "write.csv(Embeddings(seur, reduction='tsne'), file = \"tSNECoordinates.csv\")\n",
    "seur <- RunUMAP(seur, dims=1:numPCs)\n",
    "write.csv(Embeddings(seur, reduction='umap'), file = \"UMAPCoordinates.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d705cc5-f727-4768-894b-c2a66c861365",
   "metadata": {},
   "outputs": [],
   "source": [
    "runRes <- c()\n",
    "tsnePlots <- list()\n",
    "umapPlots <- list()\n",
    "for (res in resolutions) {\n",
    "  seur <- FindClusters(seur, dims=1:numPCs, resolution = res, print.output = 0, save.SNN = T)\n",
    "  tsne <- TSNEPlot(seur) + ggtitle(paste(numPCs,\"PCs_res\", res, sep=\"\")) +\n",
    "    theme(plot.title = element_text(hjust = 0.5))\n",
    "  tsnePlots[[as.character(res)]] <- tsne\n",
    "  png(paste(\"TSNEPlotwith\",numPCs,\"PCs_\", res, \".png\", sep=\"\"), height=7, width=7, units='in', res=300)\n",
    "  print(tsne)\n",
    "  dev.off()\n",
    "  umap <- DimPlot(seur, reduction=\"umap\") + ggtitle(paste(numPCs,\"PCs_res\", res, sep=\"\")) +\n",
    "    theme(plot.title = element_text(hjust = 0.5))\n",
    "  umapPlots[[as.character(res)]] <- umap\n",
    "  png(paste(\"UMAPPlotwith\",numPCs,\"PCs_\", res, \".png\", sep=\"\"), height=7, width=7, units='in', res=300)\n",
    "  print(umap)\n",
    "  dev.off()\n",
    "\n",
    "  try({\n",
    "    seur.markers <- FindAllMarkers(object = seur, logfc.threshold = 0.25, only.pos=TRUE)\n",
    "    write.csv(seur.markers %>% group_by(cluster) %>% top_n(-100,\n",
    "                                                           p_val), paste(\"top100markers_pc\", numPCs, \"_res\", res, \".csv\", sep = \"\"))\n",
    "    saveRDS(seur.markers, paste(\"markers_res\", res, \".rds\", sep = \"\"))\n",
    "    runRes <- append(runRes, res)})\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8821c18-b955-4700-923d-1249fb550764",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save object\n",
    "saveRDS(seur, file = \"seur_10x_cluster_object.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "289836d2-00b6-474a-86a6-cd4dedb774d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"TSNEPlots.pdf\")\n",
    "for (res in tsnePlots){\n",
    "  print(res)\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46b93064-dcc1-49bc-8950-46feeb0d07b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"UMAPPlots.pdf\")\n",
    "for (res in umapPlots){\n",
    "  print(res)\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01fdc8f6-9ad8-40e5-b6ca-85ff18d7dda2",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Create Silhoutte Plots\n",
    "for (res in runRes){\n",
    "  coord <- Embeddings(seur, reduction.type='pca')[,1:numPCs]\n",
    "  Idents(seur) <- seur@meta.data[[paste0('SCT_snn_res.', res)]]\n",
    "  clusters <- Idents(seur)\n",
    "  d <- dist(coord, method=\"euclidean\")\n",
    "  sil<-silhouette(as.numeric(clusters), dist=d)\n",
    "  #silPlot <- recordPlot()\n",
    "  pdf(paste0(\"SilhouettePlot_res\",res,\".pdf\"))#, height=7, width=7, units='in', res=300)\n",
    "  plot(sil, col=as.factor(clusters[order(clusters, decreasing=FALSE)]), main=paste(\"Silhouette plot of Seurat clustering - resolution \", res, sep=\"\"), lty=2)\n",
    "  abline(v=mean(sil[,3]), col=\"red4\", lty=2)\n",
    "  dev.off()\n",
    "}\n",
    "\n",
    "##Remove resolutions that failed marker generation\n",
    "print(runRes)\n",
    "for (res in setdiff(resolutions, runRes)){\n",
    "  seur@meta.data[paste0('SCT_snn_res.', res)] <- NULL\n",
    "}\n",
    "\n",
    "write(min(runRes), \"minRes.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdc254a4-7f34-47b5-8fb4-e48b75be85c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af756324-3533-445b-ab14-b1b3952b6b84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'jupyter nbconvert --to script --output sc_seurat /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.ipynb&gt; /dev/null 2&gt;&amp;1 '"
      ],
      "text/latex": [
       "'jupyter nbconvert --to script --output sc\\_seurat /mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna/sc\\_seurat.ipynb> /dev/null 2>\\&1 '"
      ],
      "text/markdown": [
       "'jupyter nbconvert --to script --output sc_seurat /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.ipynb&gt; /dev/null 2&gt;&amp;1 '"
      ],
      "text/plain": [
       "[1] \"jupyter nbconvert --to script --output sc_seurat /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.ipynb> /dev/null 2>&1 \""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "notebook_prefix = \"sc_seurat\"                                                         # %exclude_jupyterlab%\n",
    "notebook_name = paste0(notebook_prefix, \".ipynb\")                                     # %exclude_jupyterlab%\n",
    "notebook_r = paste0(script_path, \"/\", paste0(notebook_prefix, \".r\"))                  # %exclude_jupyterlab%\n",
    "notebook_path = paste0(script_path, \"/\", notebook_name)                               # %exclude_jupyterlab%\n",
    "opt_name = paste0(script_path, \"/\", sub(\".ipynb\", \"_opt.R\", notebook_name))           # %exclude_jupyterlab%\n",
    "output = paste0(script_path, \"/\", sub(\".ipynb\", \".prod.R\", notebook_name))            # %exclude_jupyterlab%\n",
    "cmd1 = paste0(\"jupyter nbconvert --to script --output \",                              # %exclude_jupyterlab%\n",
    "             notebook_prefix, ' ', notebook_path, \"> /dev/null 2>&1 \")                # %exclude_jupyterlab%\n",
    "cmd1                                                                                  # %exclude_jupyterlab%\n",
    "system(cmd1, intern = TRUE)                                                           # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d1c24963-ff07-4aa9-abe9-cea08fc9e370",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'cat -s /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat_opt.R /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.r |grep -v exclude_jupyterlab &gt; /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.prod.R 2&gt;&amp;1'"
      ],
      "text/latex": [
       "'cat -s /mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna/sc\\_seurat\\_opt.R /mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna/sc\\_seurat.r \\textbar{}grep -v exclude\\_jupyterlab > /mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna/sc\\_seurat.prod.R 2>\\&1'"
      ],
      "text/markdown": [
       "'cat -s /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat_opt.R /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.r |grep -v exclude_jupyterlab &gt; /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.prod.R 2&gt;&amp;1'"
      ],
      "text/plain": [
       "[1] \"cat -s /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat_opt.R /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.r |grep -v exclude_jupyterlab > /mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna/sc_seurat.prod.R 2>&1\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cmd2 = paste0('cat -s ', opt_name, ' ', notebook_r,                                   # %exclude_jupyterlab%\n",
    "             ' |grep -v exclude_jupyterlab > ', output,  ' 2>&1')                     # %exclude_jupyterlab%\n",
    "cmd2                                                                                  # %exclude_jupyterlab%\n",
    "system(cmd2, intern = T)                                                              # %exclude_jupyterlab%\n",
    "system(paste0(\"rm \", notebook_r))                                                     # %exclude_jupyterlab%  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85e87b25-4da6-4829-9d2f-0932ca816848",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
