{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b74265ec-5a3b-4f8b-8bc3-dc7a0748358d",
   "metadata": {},
   "outputs": [],
   "source": [
    "script_path = getwd()                                # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5ca28816-f617-4fd1-941b-ae1c6e7ecd7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna'"
      ],
      "text/latex": [
       "'/mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/scripts/rna'"
      ],
      "text/markdown": [
       "'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna'"
      ],
      "text/plain": [
       "[1] \"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/scripts/rna\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "script_path                                          # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a69fbdf1-1fe7-4f89-90ec-4fb7788ba37f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "#         '/sc_singleR_opt.R --genome=\"hg38\" --markerList=\"/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv\" --outdir=\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR\" --rds=\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/test_dir/seur_10x_cluster_object.rds\"'),  # %exclude_jupyterlab%\n",
    "#        intern = T) # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "80169370-d3b7-4432-bafa-57468beaf407",
   "metadata": {},
   "outputs": [],
   "source": [
    "# system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "#         '/sc_singleR_opt.R --genome=\"hg38\" --markerList=\"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/data/marker_genelist/human_gene_list.csv\" --outdir=\"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/test_singleR/\" --rds=\"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/split_pipe_comb/1V232R/seurat/seur_10x_cluster_object.rds\"'),  # %exclude_jupyterlab%\n",
    "#        intern = T) # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c6ce7d8d-48e0-4443-b044-2befc07018c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "        '/sc_singleR_opt.R --genome=\"hg38\" --markerList=\"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/data/marker_genelist/human_gene_list.csv\" --outdir=\"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/test_singleR/\" --rds=\"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039851_16sclibParse_080625/Analysis/split_pipe_comb/1V231R/seurat/seur_10x_cluster_object.rds\"'),  # %exclude_jupyterlab%\n",
    "       intern = T) # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c282e671-2d4d-427e-8ff2-458377a682cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "opt = readRDS(\"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/test_singleR/opt.rds\")         # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "69ba2990-065f-4dd5-95c0-a66273183bcf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$rds</dt>\n",
       "\t\t<dd>'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039851_16sclibParse_080625/Analysis/split_pipe_comb/1V231R/seurat/seur_10x_cluster_object.rds'</dd>\n",
       "\t<dt>$sampleid</dt>\n",
       "\t\t<dd>'scRNA'</dd>\n",
       "\t<dt>$genome</dt>\n",
       "\t\t<dd>'hg38'</dd>\n",
       "\t<dt>$markerList</dt>\n",
       "\t\t<dd>'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/data/marker_genelist/human_gene_list.csv'</dd>\n",
       "\t<dt>$outdir</dt>\n",
       "\t\t<dd>'/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/test_singleR/'</dd>\n",
       "\t<dt>$help</dt>\n",
       "\t\t<dd>FALSE</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$rds] '/mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039851\\_16sclibParse\\_080625/Analysis/split\\_pipe\\_comb/1V231R/seurat/seur\\_10x\\_cluster\\_object.rds'\n",
       "\\item[\\$sampleid] 'scRNA'\n",
       "\\item[\\$genome] 'hg38'\n",
       "\\item[\\$markerList] '/mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/workflow/data/marker\\_genelist/human\\_gene\\_list.csv'\n",
       "\\item[\\$outdir] '/mnt/ccrsf-static/singlecell\\_projects/CarolinaBarillasMury\\_CS039906\\_16Parselib\\_081125/Analysis/test\\_singleR/'\n",
       "\\item[\\$help] FALSE\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$rds\n",
       ":   '/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039851_16sclibParse_080625/Analysis/split_pipe_comb/1V231R/seurat/seur_10x_cluster_object.rds'\n",
       "$sampleid\n",
       ":   'scRNA'\n",
       "$genome\n",
       ":   'hg38'\n",
       "$markerList\n",
       ":   '/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/data/marker_genelist/human_gene_list.csv'\n",
       "$outdir\n",
       ":   '/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/test_singleR/'\n",
       "$help\n",
       ":   FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$rds\n",
       "[1] \"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039851_16sclibParse_080625/Analysis/split_pipe_comb/1V231R/seurat/seur_10x_cluster_object.rds\"\n",
       "\n",
       "$sampleid\n",
       "[1] \"scRNA\"\n",
       "\n",
       "$genome\n",
       "[1] \"hg38\"\n",
       "\n",
       "$markerList\n",
       "[1] \"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/workflow/data/marker_genelist/human_gene_list.csv\"\n",
       "\n",
       "$outdir\n",
       "[1] \"/mnt/ccrsf-static/singlecell_projects/CarolinaBarillasMury_CS039906_16Parselib_081125/Analysis/test_singleR/\"\n",
       "\n",
       "$help\n",
       "[1] FALSE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "opt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3ef46d0a-deaf-4956-bd73-df5e4ee1145f",
   "metadata": {},
   "outputs": [],
   "source": [
    "system(paste0(\"mkdir -p \", opt$outdir))\n",
    "setwd(opt$outdir) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f2ac430e-9d69-4141-8602-91014330ee8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘celldex’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:SingleR’:\n",
      "\n",
      "    BlueprintEncodeData, DatabaseImmuneCellExpressionData,\n",
      "    HumanPrimaryCellAtlasData, ImmGenData, MonacoImmuneData,\n",
      "    MouseRNAseqData, NovershternHematopoieticData\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(SingleR)\n",
    "library(celldex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "93f18a9c-afec-4f03-bd82-141e062ae9ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- readRDS(opt$rds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0c2e12ca-eb39-475c-b5c2-cb3b0066e816",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Layer ‘data’ is empty”\n",
      "Warning message:\n",
      "“Layer ‘scale.data’ is empty”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>30</li><li>473</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 30\n",
       "\\item 473\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 30\n",
       "2. 473\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]  30 473"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DefaultAssay(seur) = \"RNA\"\n",
    "sce <- as.SingleCellExperiment(seur)\n",
    "range(colSums(counts(sce)))\n",
    "DefaultAssay(seur) = \"SCT\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "247618ef-3e52-4bb5-9a8e-58fd83842ad0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d3f292f2-fc85-4617-b5d8-f412c1e02102",
   "metadata": {},
   "outputs": [],
   "source": [
    "sce <- sce[, colSums(counts(sce)) > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "65a5c0bf-909e-4c82-b0b7-07568642186b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>30</li><li>473</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 30\n",
       "\\item 473\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 30\n",
       "2. 473\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]  30 473"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "range(colSums(counts(sce)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "997ac054-5ea9-4b51-af81-770dc7a65a20",
   "metadata": {},
   "outputs": [],
   "source": [
    "sce <- scater::logNormCounts(sce)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1c91e550-c536-45fb-9995-8cb33474f36e",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneList <- read.csv(opt$markerList, header=FALSE, row.names=1, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7fc08dcd-13d0-4a08-a7a8-2c0df3def405",
   "metadata": {},
   "outputs": [],
   "source": [
    "## plotScoreDistribution: use this plot to check that outlier detection in pruneScores() behaved sensibly. \n",
    "## Labels with especially low deltas may warrant some additional caution in their interpretation.\n",
    "generatePlots <- function(name, pred, seur, markers){\n",
    "  saveRDS(pred, paste0('pred_', name, '.rds'))\n",
    "  #pdf(paste0('heatmap_', name, '.pdf'))\n",
    "  #print(plotScoreHeatmap(pred, clusters=sce$res.0.1))\n",
    "  #dev.off()\n",
    "  pdf(paste0('ScoreDistribution_', name, '.pdf'), height=15)\n",
    "  # plotScoreDistribution deprecated\n",
    "  #print(plotScoreDistribution(pred, show = \"delta.med\", ncol = 3, show.nmads = 3))\n",
    "  tryCatch({\n",
    "    p <- plotDeltaDistribution(pred, show=\"delta.med\", ncol=3)\n",
    "    print(p)\n",
    "  }, error=function(e) message(\"Skipping delta distribution plot: \", conditionMessage(e)))\n",
    "  dev.off()\n",
    "  seur@meta.data[name] <- ''\n",
    "  seur@meta.data[rownames(pred),name] <- pred$pruned.labels\n",
    "  Idents(seur) <- seur@meta.data[,name]\n",
    "  png(paste0(\"UMAP_\", name, \".png\"))\n",
    "  print(DimPlot(seur))\n",
    "  dev.off()\n",
    "  png(paste0(\"TSNE_\", name, \".png\"))\n",
    "  print(DimPlot(seur, reduction='tsne'))\n",
    "  dev.off()\n",
    "  dir.create(\"gene_list_plots\")\n",
    "  for (i in 1:dim(markers)[1]) {\n",
    "    genes.use <- as.character(markers[row.names(markers)[i],])\n",
    "    numGenes <- length(genes.use[genes.use %in% row.names(seur)])\n",
    "    if (sum(genes.use %in% row.names(seur)) > 1) {\n",
    "      png(paste0('gene_list_plots/ViolinPlot_', name, '_', row.names(geneList)[i],'.png'), height = 5, width = (length(genes.use[genes.use %in% row.names(seur)])+1)*5, units='in', res=300)\n",
    "      print(suppressWarnings(VlnPlot(seur, genes.use[genes.use %in% row.names(seur)], ncol=numGenes)))\n",
    "      dev.off()\n",
    "    }\n",
    "  }\n",
    "  return(seur)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aa3997e2-b636-4dec-b5a8-2d708a59064b",
   "metadata": {},
   "outputs": [],
   "source": [
    "generateMarkerPlots <- function(seur, markers) {\n",
    "  dir.create(\"gene_list_plots\")\n",
    "  for (i in 1:dim(markers)[1]) {\n",
    "    genes.use <- as.character(markers[row.names(markers)[i],])\n",
    "    numGenes <- length(genes.use[genes.use %in% row.names(seur)])\n",
    "    if (sum(genes.use %in% row.names(seur)) > 1) {\n",
    "      png(paste0('gene_list_plots/FeaturePlot_', row.names(geneList)[i],'.png'), height = 5, width = (length(genes.use[genes.use %in% row.names(seur)])+1)*5, units='in', res=300)\n",
    "      print(suppressWarnings(FeaturePlot(seur, genes.use[genes.use %in% row.names(seur)], ncol=numGenes)))\n",
    "      dev.off()\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0d9bd8a-30a8-4f2b-8293-37c97498b03c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The objects returned by functions:  ImmGenData, MouseRNAseqData, \n",
    "# HumanPrimaryCellAtlasData and BlueprintEncodeData are saved under \n",
    "# /home/docker/. So you only need to load the RDS files directly. \n",
    "if (startsWith(opt$genome, \"mm10\") | startsWith(opt$genome, \"mm39\")) {\n",
    "  #immgen <- ImmGenData()# - mouse\n",
    "  #mouserna <- MouseRNAseqData()# - mouse\n",
    "  immgen <- readRDS(\"/home/docker/ImmGenData.rds\")\n",
    "  mouserna <- readRDS(\"/home/docker/MouseRNAseqData.rds\")\n",
    "  pred.multi <- SingleR(test = sce,\n",
    "      ref = list(IG=immgen, MRNA=mouserna),\n",
    "      labels = list(immgen$label.main, mouserna$label.main))\n",
    "  seur <- generatePlots('immgen_mourserna', pred.multi, seur, geneList)\n",
    "  temp <- table(pred.multi$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_Immgen-MouseRNA_general.csv', row.names=FALSE)\n",
    "\n",
    "  pred.immgen <- SingleR(test = sce,\n",
    "      ref = immgen, labels = immgen$label.main)\n",
    "  seur <- generatePlots('immgen', pred.immgen, seur, geneList)\n",
    "  temp <- table(pred.immgen$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_Immgen_general.csv', row.names=FALSE)\n",
    "\n",
    "  pred.mouserna <- SingleR(test = sce,\n",
    "      ref = mouserna, labels = mouserna$label.main)\n",
    "  seur <- generatePlots('mourserna', pred.mouserna, seur, geneList)\n",
    "  temp <- table(pred.mouserna$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_MouseRNA_general.csv', row.names=FALSE)\n",
    "\n",
    "  write.csv(seur@meta.data[,c('immgen_mourserna', 'immgen', 'mourserna')], 'annotations_compiled_barcode.csv', quote=FALSE)\n",
    "} else if (startsWith(opt$genome, \"hg38\") | startsWith(opt$genome, \"hg19\")) {\n",
    "  #hpca <- HumanPrimaryCellAtlasData()\n",
    "  #blueprint <- BlueprintEncodeData()\n",
    "  hpca <- readRDS(\"/home/docker/HumanPrimaryCellAtlasData.rds\")\n",
    "  blueprint <- readRDS(\"/home/docker/BlueprintEncodeData.rds\")\n",
    "  pred.multi <- SingleR(test = sce,\n",
    "      ref = list(BP=blueprint, HPCA=hpca),\n",
    "      labels = list(blueprint$label.main, hpca$label.main))\n",
    "  seur <- generatePlots('blueprintencode_hpca', pred.multi, seur, geneList)\n",
    "  temp <- table(pred.multi$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_BlueprintENCODE-HPCA_general.csv', row.names=FALSE)\n",
    "\n",
    "  pred.blueprint <- SingleR(test = sce,\n",
    "      ref = blueprint, labels = blueprint$label.main)\n",
    "  seur <- generatePlots('blueprintencode', pred.blueprint, seur, geneList)\n",
    "  temp <- table(pred.blueprint$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_BlueprintENCODE_general.csv', row.names=FALSE)\n",
    "\n",
    "  pred.hpca <- SingleR(test = sce,\n",
    "      ref = hpca, labels = hpca$label.main)\n",
    "  seur <- generatePlots('hpca', pred.hpca, seur, geneList)\n",
    "  temp <- table(pred.hpca$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_HPCA_general.csv', row.names=FALSE)\n",
    "\n",
    "  write.csv(seur@meta.data[,c('blueprintencode_hpca', 'blueprintencode', 'hpca')], 'annotations_compiled_barcode.csv', quote=FALSE)\n",
    "}\n",
    "\n",
    "suppressWarnings(generateMarkerPlots(seur, geneList))\n",
    "saveRDS(seur, 'seur_10x_cluster_singler.rds')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdc254a4-7f34-47b5-8fb4-e48b75be85c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af756324-3533-445b-ab14-b1b3952b6b84",
   "metadata": {},
   "outputs": [],
   "source": [
    "notebook_prefix = \"sc_singleR\"                                                        # %exclude_jupyterlab%\n",
    "notebook_name = paste0(notebook_prefix, \".ipynb\")                                     # %exclude_jupyterlab%\n",
    "notebook_r = paste0(script_path, \"/\", paste0(notebook_prefix, \".r\"))                  # %exclude_jupyterlab%\n",
    "notebook_path = paste0(script_path, \"/\", notebook_name)                               # %exclude_jupyterlab%\n",
    "opt_name = paste0(script_path, \"/\", sub(\".ipynb\", \"_opt.R\", notebook_name))           # %exclude_jupyterlab%\n",
    "output = paste0(script_path, \"/\", sub(\".ipynb\", \".prod.R\", notebook_name))            # %exclude_jupyterlab%\n",
    "cmd1 = paste0(\"jupyter nbconvert --to script --output \",                              # %exclude_jupyterlab%\n",
    "             notebook_prefix, ' ', notebook_path, \"> /dev/null 2>&1 \")                # %exclude_jupyterlab%\n",
    "cmd1                                                                                  # %exclude_jupyterlab%\n",
    "system(cmd1, intern = TRUE)                                                           # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1c24963-ff07-4aa9-abe9-cea08fc9e370",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd2 = paste0('cat -s ', opt_name, ' ', notebook_r,                                   # %exclude_jupyterlab%\n",
    "             ' |grep -v exclude_jupyterlab > ', output,  ' 2>&1')                     # %exclude_jupyterlab%\n",
    "cmd2                                                                                  # %exclude_jupyterlab%\n",
    "system(cmd2, intern = T)                                                              # %exclude_jupyterlab%\n",
    "system(paste0(\"rm \", notebook_r))                                                     # %exclude_jupyterlab%  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85e87b25-4da6-4829-9d2f-0932ca816848",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ededa7dc-5de8-44bb-8789-33b08134a04e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
