{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b74265ec-5a3b-4f8b-8bc3-dc7a0748358d",
   "metadata": {},
   "outputs": [],
   "source": [
    "script_path = getwd()                                # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ca28816-f617-4fd1-941b-ae1c6e7ecd7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna'"
      ],
      "text/latex": [
       "'/Volumes/ccrsf-ifx/Software/github/SF\\_scMaestro/workflow/scripts/rna'"
      ],
      "text/markdown": [
       "'/Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna'"
      ],
      "text/plain": [
       "[1] \"/Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "script_path                                          # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a69fbdf1-1fe7-4f89-90ec-4fb7788ba37f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in paste0(\"Rscript \", script_path, \"/sc_singleR_opt.R --genome=\\\"hg38\\\" --markerList=\\\"/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv\\\" --outdir=\\\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR\\\" --rds=\\\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/test_dir/seur_10x_cluster_object.rds\\\"\"): object 'script_path' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in paste0(\"Rscript \", script_path, \"/sc_singleR_opt.R --genome=\\\"hg38\\\" --markerList=\\\"/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv\\\" --outdir=\\\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR\\\" --rds=\\\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/test_dir/seur_10x_cluster_object.rds\\\"\"): object 'script_path' not found\nTraceback:\n",
      "1. system(paste0(\"Rscript \", script_path, \"/sc_singleR_opt.R --genome=\\\"hg38\\\" --markerList=\\\"/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv\\\" --outdir=\\\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR\\\" --rds=\\\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/test_dir/seur_10x_cluster_object.rds\\\"\"), \n .     intern = T)",
      "2. paste0(\"Rscript \", script_path, \"/sc_singleR_opt.R --genome=\\\"hg38\\\" --markerList=\\\"/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv\\\" --outdir=\\\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR\\\" --rds=\\\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/test_dir/seur_10x_cluster_object.rds\\\"\")"
     ]
    }
   ],
   "source": [
    "system(paste0('Rscript ', script_path,                # %exclude_jupyterlab%\n",
    "        '/sc_singleR_opt.R --genome=\"hg38\" --markerList=\"/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv\" --outdir=\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR\" --rds=\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/test_dir/seur_10x_cluster_object.rds\"'),  # %exclude_jupyterlab%\n",
    "       intern = T) # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c282e671-2d4d-427e-8ff2-458377a682cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "opt = readRDS(\"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR/opt.rds\")         # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "69ba2990-065f-4dd5-95c0-a66273183bcf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$rds</dt>\n",
       "\t\t<dd>'/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/test_dir/seur_10x_cluster_object.rds'</dd>\n",
       "\t<dt>$sampleid</dt>\n",
       "\t\t<dd>'scRNA'</dd>\n",
       "\t<dt>$genome</dt>\n",
       "\t\t<dd>'hg38'</dd>\n",
       "\t<dt>$markerList</dt>\n",
       "\t\t<dd>'/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv'</dd>\n",
       "\t<dt>$outdir</dt>\n",
       "\t\t<dd>'/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR'</dd>\n",
       "\t<dt>$help</dt>\n",
       "\t\t<dd>FALSE</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$rds] '/Volumes/ccrsf-static/Analysis/xies4/github\\_repos/pipeline\\_dev\\_test/test\\_dir/seur\\_10x\\_cluster\\_object.rds'\n",
       "\\item[\\$sampleid] 'scRNA'\n",
       "\\item[\\$genome] 'hg38'\n",
       "\\item[\\$markerList] '/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single\\_cell/gene\\_lists/human\\_gene\\_list.csv'\n",
       "\\item[\\$outdir] '/Volumes/ccrsf-static/Analysis/xies4/github\\_repos/pipeline\\_dev\\_test/singleR'\n",
       "\\item[\\$help] FALSE\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$rds\n",
       ":   '/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/test_dir/seur_10x_cluster_object.rds'\n",
       "$sampleid\n",
       ":   'scRNA'\n",
       "$genome\n",
       ":   'hg38'\n",
       "$markerList\n",
       ":   '/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv'\n",
       "$outdir\n",
       ":   '/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR'\n",
       "$help\n",
       ":   FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$rds\n",
       "[1] \"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/test_dir/seur_10x_cluster_object.rds\"\n",
       "\n",
       "$sampleid\n",
       "[1] \"scRNA\"\n",
       "\n",
       "$genome\n",
       "[1] \"hg38\"\n",
       "\n",
       "$markerList\n",
       "[1] \"/Volumes/ccrsf-ifx/Software/scripts/bin/currentsnake/single_cell/gene_lists/human_gene_list.csv\"\n",
       "\n",
       "$outdir\n",
       "[1] \"/Volumes/ccrsf-static/Analysis/xies4/github_repos/pipeline_dev_test/singleR\"\n",
       "\n",
       "$help\n",
       "[1] FALSE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "opt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ef46d0a-deaf-4956-bd73-df5e4ee1145f",
   "metadata": {},
   "outputs": [],
   "source": [
    "system(paste0(\"mkdir -p \", opt$outdir))\n",
    "setwd(opt$outdir) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f2ac430e-9d69-4141-8602-91014330ee8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘celldex’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:SingleR’:\n",
      "\n",
      "    BlueprintEncodeData, DatabaseImmuneCellExpressionData,\n",
      "    HumanPrimaryCellAtlasData, ImmGenData, MonacoImmuneData,\n",
      "    MouseRNAseqData, NovershternHematopoieticData\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(SingleR)\n",
    "library(celldex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "93f18a9c-afec-4f03-bd82-141e062ae9ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "seur <- readRDS(opt$rds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0c2e12ca-eb39-475c-b5c2-cb3b0066e816",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Layer ‘data’ is empty”\n",
      "Warning message:\n",
      "“Layer ‘scale.data’ is empty”\n"
     ]
    }
   ],
   "source": [
    "sce <- as.SingleCellExperiment(seur)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "997ac054-5ea9-4b51-af81-770dc7a65a20",
   "metadata": {},
   "outputs": [],
   "source": [
    "sce <- scater::logNormCounts(sce)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1c91e550-c536-45fb-9995-8cb33474f36e",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneList <- read.csv(opt$markerList, header=FALSE, row.names=1, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7fc08dcd-13d0-4a08-a7a8-2c0df3def405",
   "metadata": {},
   "outputs": [],
   "source": [
    "## plotScoreDistribution: use this plot to check that outlier detection in pruneScores() behaved sensibly. \n",
    "## Labels with especially low deltas may warrant some additional caution in their interpretation.\n",
    "generatePlots <- function(name, pred, seur, markers){\n",
    "  saveRDS(pred, paste0('pred_', name, '.rds'))\n",
    "  #pdf(paste0('heatmap_', name, '.pdf'))\n",
    "  #print(plotScoreHeatmap(pred, clusters=sce$res.0.1))\n",
    "  #dev.off()\n",
    "  pdf(paste0('ScoreDistribution_', name, '.pdf'), height=15)\n",
    "  # plotScoreDistribution deprecated\n",
    "  #print(plotScoreDistribution(pred, show = \"delta.med\", ncol = 3, show.nmads = 3))\n",
    "  print(plotDeltaDistribution(pred, show = \"delta.med\", ncol = 3))\n",
    "  dev.off()\n",
    "  seur@meta.data[name] <- ''\n",
    "  seur@meta.data[rownames(pred),name] <- pred$pruned.labels\n",
    "  Idents(seur) <- seur@meta.data[,name]\n",
    "  pdf(paste0(\"UMAP_\", name, \".pdf\"))\n",
    "  print(DimPlot(seur))\n",
    "  dev.off()\n",
    "  pdf(paste0(\"TSNE_\", name, \".pdf\"))\n",
    "  print(DimPlot(seur, reduction='tsne'))\n",
    "  dev.off()\n",
    "  dir.create(\"gene_list_plots\")\n",
    "  for (i in 1:dim(markers)[1]) {\n",
    "    genes.use <- as.character(markers[row.names(markers)[i],])\n",
    "    numGenes <- length(genes.use[genes.use %in% row.names(seur)])\n",
    "    if (sum(genes.use %in% row.names(seur)) > 1) {\n",
    "      png(paste0('gene_list_plots/ViolinPlot_', name, '_', row.names(geneList)[i],'.png'), height = 5, width = (length(genes.use[genes.use %in% row.names(seur)])+1)*5, units='in', res=300)\n",
    "      print(suppressWarnings(VlnPlot(seur, genes.use[genes.use %in% row.names(seur)], ncol=numGenes)))\n",
    "      dev.off()\n",
    "    }\n",
    "  }\n",
    "  return(seur)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "aa3997e2-b636-4dec-b5a8-2d708a59064b",
   "metadata": {},
   "outputs": [],
   "source": [
    "generateMarkerPlots <- function(seur, markers) {\n",
    "  dir.create(\"gene_list_plots\")\n",
    "  for (i in 1:dim(markers)[1]) {\n",
    "    genes.use <- as.character(markers[row.names(markers)[i],])\n",
    "    numGenes <- length(genes.use[genes.use %in% row.names(seur)])\n",
    "    if (sum(genes.use %in% row.names(seur)) > 1) {\n",
    "      png(paste0('gene_list_plots/FeaturePlot_', row.names(geneList)[i],'.png'), height = 5, width = (length(genes.use[genes.use %in% row.names(seur)])+1)*5, units='in', res=300)\n",
    "      print(suppressWarnings(FeaturePlot(seur, genes.use[genes.use %in% row.names(seur)], ncol=numGenes)))\n",
    "      dev.off()\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a0d9bd8a-30a8-4f2b-8293-37c97498b03c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message in max(data$density, na.rm = TRUE):\n",
      "“no non-missing arguments to max; returning -Inf”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mComputation failed in `stat_ydensity()`.\n",
      "Caused by error in `$<-.data.frame`:\n",
      "\u001b[33m!\u001b[39m replacement has 1 row, data has 0”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message in max(data$density, na.rm = TRUE):\n",
      "“no non-missing arguments to max; returning -Inf”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mComputation failed in `stat_ydensity()`.\n",
      "Caused by error in `$<-.data.frame`:\n",
      "\u001b[33m!\u001b[39m replacement has 1 row, data has 0”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TableGrob (2 x 1) \"arrange\": 2 grobs\n",
      "  z     cells    name           grob\n",
      "1 1 (1-1,1-1) arrange gtable[layout]\n",
      "2 2 (2-2,1-1) arrange gtable[layout]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(\"gene_list_plots\"):\n",
      "“'gene_list_plots' already exists”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message in max(data$density, na.rm = TRUE):\n",
      "“no non-missing arguments to max; returning -Inf”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mComputation failed in `stat_ydensity()`.\n",
      "Caused by error in `$<-.data.frame`:\n",
      "\u001b[33m!\u001b[39m replacement has 1 row, data has 0”\n",
      "Warning message in dir.create(\"gene_list_plots\"):\n",
      "“'gene_list_plots' already exists”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message in max(data$density, na.rm = TRUE):\n",
      "“no non-missing arguments to max; returning -Inf”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mComputation failed in `stat_ydensity()`.\n",
      "Caused by error in `$<-.data.frame`:\n",
      "\u001b[33m!\u001b[39m replacement has 1 row, data has 0”\n",
      "Warning message in dir.create(\"gene_list_plots\"):\n",
      "“'gene_list_plots' already exists”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mGroups with fewer than two datapoints have been dropped.\n",
      "\u001b[36mℹ\u001b[39m Set `drop = FALSE` to consider such groups for position adjustment purposes.”\n"
     ]
    }
   ],
   "source": [
    "# The objects returned by functions:  ImmGenData, MouseRNAseqData, \n",
    "# HumanPrimaryCellAtlasData and BlueprintEncodeData are saved under \n",
    "# /home/docker/. So you only need to load the RDS files directly. \n",
    "if (opt$genome == \"mm10\") {\n",
    "  #immgen <- ImmGenData()# - mouse\n",
    "  #mouserna <- MouseRNAseqData()# - mouse\n",
    "  immgen <- readRDS(\"/home/docker/ImmGenData.rds\")\n",
    "  mouserna <- readRDS(\"/home/docker/MouseRNAseqData.rds\")\n",
    "  pred.multi <- SingleR(test = sce,\n",
    "      ref = list(IG=immgen, MRNA=mouserna),\n",
    "      labels = list(immgen$label.main, mouserna$label.main))\n",
    "  seur <- generatePlots('immgen_mourserna', pred.multi, seur, geneList)\n",
    "  temp <- table(pred.multi$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_Immgen-MouseRNA_general.csv', row.names=FALSE)\n",
    "\n",
    "  pred.immgen <- SingleR(test = sce,\n",
    "      ref = immgen, labels = immgen$label.main)\n",
    "  seur <- generatePlots('immgen', pred.immgen, seur, geneList)\n",
    "  temp <- table(pred.immgen$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_Immgen_general.csv', row.names=FALSE)\n",
    "\n",
    "  pred.mouserna <- SingleR(test = sce,\n",
    "      ref = mouserna, labels = mouserna$label.main)\n",
    "  seur <- generatePlots('mourserna', pred.mouserna, seur, geneList)\n",
    "  temp <- table(pred.mouserna$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_MouseRNA_general.csv', row.names=FALSE)\n",
    "\n",
    "  write.csv(seur@meta.data[,c('immgen_mourserna', 'immgen', 'mourserna')], 'annotations_compiled_barcode.csv', quote=FALSE)\n",
    "} else if (opt$genome == \"hg38\") {\n",
    "  #hpca <- HumanPrimaryCellAtlasData()\n",
    "  #blueprint <- BlueprintEncodeData()\n",
    "  hpca <- readRDS(\"/home/docker/HumanPrimaryCellAtlasData.rds\")\n",
    "  blueprint <- readRDS(\"/home/docker/BlueprintEncodeData.rds\")\n",
    "  pred.multi <- SingleR(test = sce,\n",
    "      ref = list(BP=blueprint, HPCA=hpca),\n",
    "      labels = list(blueprint$label.main, hpca$label.main))\n",
    "  seur <- generatePlots('blueprintencode_hpca', pred.multi, seur, geneList)\n",
    "  temp <- table(pred.multi$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_BlueprintENCODE-HPCA_general.csv', row.names=FALSE)\n",
    "\n",
    "  pred.blueprint <- SingleR(test = sce,\n",
    "      ref = blueprint, labels = blueprint$label.main)\n",
    "  seur <- generatePlots('blueprintencode', pred.blueprint, seur, geneList)\n",
    "  temp <- table(pred.blueprint$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_BlueprintENCODE_general.csv', row.names=FALSE)\n",
    "\n",
    "  pred.hpca <- SingleR(test = sce,\n",
    "      ref = hpca, labels = hpca$label.main)\n",
    "  seur <- generatePlots('hpca', pred.hpca, seur, geneList)\n",
    "  temp <- table(pred.hpca$pruned.labels)\n",
    "  write.csv(sort(temp, decreasing=TRUE), 'annotations_HPCA_general.csv', row.names=FALSE)\n",
    "\n",
    "  write.csv(seur@meta.data[,c('blueprintencode_hpca', 'blueprintencode', 'hpca')], 'annotations_compiled_barcode.csv', quote=FALSE)\n",
    "}\n",
    "\n",
    "suppressWarnings(generateMarkerPlots(seur, geneList))\n",
    "saveRDS(seur, 'seur_10x_cluster_singler.rds')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bdc254a4-7f34-47b5-8fb4-e48b75be85c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-pc-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 22.04.4 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 \n",
       "LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Etc/UTC\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] celldex_1.12.0              SingleR_2.4.1              \n",
       " [3] SummarizedExperiment_1.32.0 Biobase_2.62.0             \n",
       " [5] GenomicRanges_1.54.1        GenomeInfoDb_1.38.8        \n",
       " [7] IRanges_2.37.1              S4Vectors_0.40.2           \n",
       " [9] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      \n",
       "[11] matrixStats_1.3.0           Seurat_5.0.1               \n",
       "[13] SeuratObject_5.0.1          sp_2.1-3                   \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22              splines_4.3.3                \n",
       "  [3] later_1.3.2                   pbdZMQ_0.3-11                \n",
       "  [5] filelock_1.0.3                bitops_1.0-7                 \n",
       "  [7] tibble_3.2.1                  polyclip_1.10-6              \n",
       "  [9] fastDummies_1.7.3             lifecycle_1.0.4              \n",
       " [11] globals_0.16.3                lattice_0.22-6               \n",
       " [13] MASS_7.3-60.0.1               magrittr_2.0.3               \n",
       " [15] plotly_4.10.4                 yaml_2.3.8                   \n",
       " [17] httpuv_1.6.15                 sctransform_0.4.1            \n",
       " [19] spam_2.10-0                   spatstat.sparse_3.0-3        \n",
       " [21] reticulate_1.35.0             cowplot_1.1.3                \n",
       " [23] pbapply_1.7-2                 DBI_1.2.2                    \n",
       " [25] RColorBrewer_1.1-3            abind_1.4-5                  \n",
       " [27] zlibbioc_1.48.2               Rtsne_0.17                   \n",
       " [29] purrr_1.0.2                   RCurl_1.98-1.14              \n",
       " [31] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      \n",
       " [33] ggrepel_0.9.5                 irlba_2.3.5.1                \n",
       " [35] listenv_0.9.1                 spatstat.utils_3.0-4         \n",
       " [37] goftest_1.2-3                 RSpectra_0.16-1              \n",
       " [39] spatstat.random_3.2-3         fitdistrplus_1.1-11          \n",
       " [41] parallelly_1.37.1             DelayedMatrixStats_1.24.0    \n",
       " [43] leiden_0.4.3.1                codetools_0.2-20             \n",
       " [45] DelayedArray_0.28.0           scuttle_1.12.0               \n",
       " [47] tidyselect_1.2.1              farver_2.1.1                 \n",
       " [49] viridis_0.6.5                 ScaledMatrix_1.10.0          \n",
       " [51] BiocFileCache_2.10.2          base64enc_0.1-3              \n",
       " [53] spatstat.explore_3.2-7        jsonlite_1.8.8               \n",
       " [55] BiocNeighbors_1.20.2          progressr_0.14.0             \n",
       " [57] scater_1.30.1                 ggridges_0.5.6               \n",
       " [59] survival_3.5-8                tools_4.3.3                  \n",
       " [61] ica_1.0-3                     Rcpp_1.0.12                  \n",
       " [63] glue_1.7.0                    gridExtra_2.3                \n",
       " [65] SparseArray_1.2.4             IRdisplay_1.1                \n",
       " [67] dplyr_1.1.4                   withr_3.0.0                  \n",
       " [69] BiocManager_1.30.22           fastmap_1.1.1                \n",
       " [71] fansi_1.0.6                   digest_0.6.35                \n",
       " [73] rsvd_1.0.5                    R6_2.5.1                     \n",
       " [75] mime_0.12                     colorspace_2.1-0             \n",
       " [77] scattermore_1.2               tensor_1.5                   \n",
       " [79] spatstat.data_3.0-4           RSQLite_2.3.6                \n",
       " [81] utf8_1.2.4                    tidyr_1.3.1                  \n",
       " [83] generics_0.1.3                data.table_1.15.4            \n",
       " [85] httr_1.4.7                    htmlwidgets_1.6.4            \n",
       " [87] S4Arrays_1.2.1                uwot_0.1.16                  \n",
       " [89] pkgconfig_2.0.3               gtable_0.3.4                 \n",
       " [91] blob_1.2.4                    lmtest_0.9-40                \n",
       " [93] SingleCellExperiment_1.24.0   XVector_0.42.0               \n",
       " [95] htmltools_0.5.8.1             dotCall64_1.1-1              \n",
       " [97] scales_1.3.0                  png_0.1-8                    \n",
       " [99] reshape2_1.4.4                uuid_1.2-0                   \n",
       "[101] curl_5.2.1                    nlme_3.1-164                 \n",
       "[103] repr_1.1.7                    zoo_1.8-12                   \n",
       "[105] cachem_1.0.8                  stringr_1.5.1                \n",
       "[107] BiocVersion_3.18.1            KernSmooth_2.23-22           \n",
       "[109] vipor_0.4.7                   parallel_4.3.3               \n",
       "[111] miniUI_0.1.1.1                AnnotationDbi_1.64.1         \n",
       "[113] ggrastr_1.0.2                 pillar_1.9.0                 \n",
       "[115] grid_4.3.3                    vctrs_0.6.5                  \n",
       "[117] RANN_2.6.1                    promises_1.3.0               \n",
       "[119] BiocSingular_1.18.0           dbplyr_2.5.0                 \n",
       "[121] beachmat_2.18.1               xtable_1.8-4                 \n",
       "[123] cluster_2.1.6                 beeswarm_0.4.0               \n",
       "[125] evaluate_0.23                 cli_3.6.2                    \n",
       "[127] compiler_4.3.3                rlang_1.1.3                  \n",
       "[129] crayon_1.5.2                  future.apply_1.11.2          \n",
       "[131] labeling_0.4.3                ggbeeswarm_0.7.2             \n",
       "[133] plyr_1.8.9                    stringi_1.8.3                \n",
       "[135] viridisLite_0.4.2             deldir_2.0-4                 \n",
       "[137] BiocParallel_1.36.0           munsell_0.5.1                \n",
       "[139] Biostrings_2.70.3             lazyeval_0.2.2               \n",
       "[141] spatstat.geom_3.2-9           Matrix_1.6-5                 \n",
       "[143] ExperimentHub_2.10.0          IRkernel_1.1.1               \n",
       "[145] RcppHNSW_0.6.0                patchwork_1.2.0              \n",
       "[147] sparseMatrixStats_1.14.0      bit64_4.0.5                  \n",
       "[149] future_1.33.2                 ggplot2_3.5.0                \n",
       "[151] KEGGREST_1.42.0               shiny_1.8.1.1                \n",
       "[153] interactiveDisplayBase_1.40.0 AnnotationHub_3.10.1         \n",
       "[155] ROCR_1.0-11                   igraph_2.0.3                 \n",
       "[157] memoise_2.0.1                 bit_4.0.5                    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "af756324-3533-445b-ab14-b1b3952b6b84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'jupyter nbconvert --to script --output sc_singleR /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR.ipynb&gt; /dev/null 2&gt;&amp;1 '"
      ],
      "text/latex": [
       "'jupyter nbconvert --to script --output sc\\_singleR /Volumes/ccrsf-ifx/Software/github/SF\\_scMaestro/workflow/scripts/rna/sc\\_singleR.ipynb> /dev/null 2>\\&1 '"
      ],
      "text/markdown": [
       "'jupyter nbconvert --to script --output sc_singleR /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR.ipynb&gt; /dev/null 2&gt;&amp;1 '"
      ],
      "text/plain": [
       "[1] \"jupyter nbconvert --to script --output sc_singleR /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR.ipynb> /dev/null 2>&1 \""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "notebook_prefix = \"sc_singleR\"                                                         # %exclude_jupyterlab%\n",
    "notebook_name = paste0(notebook_prefix, \".ipynb\")                                     # %exclude_jupyterlab%\n",
    "notebook_r = paste0(script_path, \"/\", paste0(notebook_prefix, \".r\"))                  # %exclude_jupyterlab%\n",
    "notebook_path = paste0(script_path, \"/\", notebook_name)                               # %exclude_jupyterlab%\n",
    "opt_name = paste0(script_path, \"/\", sub(\".ipynb\", \"_opt.R\", notebook_name))           # %exclude_jupyterlab%\n",
    "output = paste0(script_path, \"/\", sub(\".ipynb\", \".prod.R\", notebook_name))            # %exclude_jupyterlab%\n",
    "cmd1 = paste0(\"jupyter nbconvert --to script --output \",                              # %exclude_jupyterlab%\n",
    "             notebook_prefix, ' ', notebook_path, \"> /dev/null 2>&1 \")                # %exclude_jupyterlab%\n",
    "cmd1                                                                                  # %exclude_jupyterlab%\n",
    "system(cmd1, intern = TRUE)                                                            # %exclude_jupyterlab%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d1c24963-ff07-4aa9-abe9-cea08fc9e370",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'cat -s /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR_opt.R /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR.r |grep -v exclude_jupyterlab &gt; /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR.prod.R 2&gt;&amp;1'"
      ],
      "text/latex": [
       "'cat -s /Volumes/ccrsf-ifx/Software/github/SF\\_scMaestro/workflow/scripts/rna/sc\\_singleR\\_opt.R /Volumes/ccrsf-ifx/Software/github/SF\\_scMaestro/workflow/scripts/rna/sc\\_singleR.r \\textbar{}grep -v exclude\\_jupyterlab > /Volumes/ccrsf-ifx/Software/github/SF\\_scMaestro/workflow/scripts/rna/sc\\_singleR.prod.R 2>\\&1'"
      ],
      "text/markdown": [
       "'cat -s /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR_opt.R /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR.r |grep -v exclude_jupyterlab &gt; /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR.prod.R 2&gt;&amp;1'"
      ],
      "text/plain": [
       "[1] \"cat -s /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR_opt.R /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR.r |grep -v exclude_jupyterlab > /Volumes/ccrsf-ifx/Software/github/SF_scMaestro/workflow/scripts/rna/sc_singleR.prod.R 2>&1\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cmd2 = paste0('cat -s ', opt_name, ' ', notebook_r,                                      # %exclude_jupyterlab%\n",
    "             ' |grep -v exclude_jupyterlab > ', output,  ' 2>&1')                     # %exclude_jupyterlab%\n",
    "cmd2                                                                                  # %exclude_jupyterlab%\n",
    "system(cmd2, intern = T)                                                              # %exclude_jupyterlab%\n",
    "system(paste0(\"rm \", notebook_r))                                                     # %exclude_jupyterlab%  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85e87b25-4da6-4829-9d2f-0932ca816848",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
